--- a/drivers/net/ethernet/airoha/airoha_eth.c
+++ b/drivers/net/ethernet/airoha/airoha_eth.c
@@ -1046,6 +1046,15 @@ static void airoha_qdma_cleanup_rx_queue
 			FIELD_PREP(RX_RING_DMA_IDX_MASK, q->tail));
 }
 
+static void airoha_qdma_rx_queue_deinit(struct airoha_queue *q)
+{
+	airoha_qdma_cleanup_rx_queue(q);
+	if (q->page_pool) {
+		page_pool_destroy(q->page_pool);
+		q->page_pool = NULL;
+	}
+}
+
 static int airoha_qdma_init_rx(struct airoha_qdma *qdma)
 {
 	int i;
@@ -1688,11 +1697,7 @@ static void airoha_qdma_cleanup(struct a
 			continue;
 
 		netif_napi_del(&qdma->q_rx[i].napi);
-		airoha_qdma_cleanup_rx_queue(&qdma->q_rx[i]);
-		if (qdma->q_rx[i].page_pool) {
-			page_pool_destroy(qdma->q_rx[i].page_pool);
-			qdma->q_rx[i].page_pool = NULL;
-		}
+		airoha_qdma_rx_queue_deinit(&qdma->q_rx[i]);
 	}
 
 	for (i = 0; i < ARRAY_SIZE(qdma->q_tx_irq); i++) {
