--- a/drivers/net/ethernet/airoha/airoha_eth.c
+++ b/drivers/net/ethernet/airoha/airoha_eth.c
@@ -1237,7 +1237,6 @@ static int airoha_qdma_init_tx_queue(str
 {
 	struct airoha_eth *eth = qdma->eth;
 	int i, qid = q - &qdma->q_tx[0];
-	dma_addr_t dma_addr;
 
 	spin_lock_init(&q->lock);
 	q->qdma = qdma;
@@ -1250,7 +1249,7 @@ static int airoha_qdma_init_tx_queue(str
 		return -ENOMEM;
 
 	q->desc = dmam_alloc_coherent(eth->dev, size * sizeof(*q->desc),
-				      &dma_addr, GFP_KERNEL);
+				      &q->dma_addr, GFP_KERNEL);
 	if (!q->desc)
 		return -ENOMEM;
 
@@ -1266,7 +1265,7 @@ static int airoha_qdma_init_tx_queue(str
 	airoha_qdma_set(qdma, REG_TX_RING_BLOCKING(qid),
 			TX_RING_IRQ_BLOCKING_TX_DROP_EN_MASK);
 
-	airoha_qdma_wr(qdma, REG_TX_RING_BASE(qid), dma_addr);
+	airoha_qdma_wr(qdma, REG_TX_RING_BASE(qid), q->dma_addr);
 	airoha_qdma_rmw(qdma, REG_TX_CPU_IDX(qid), TX_RING_CPU_IDX_MASK,
 			FIELD_PREP(TX_RING_CPU_IDX_MASK, 0));
 	airoha_qdma_rmw(qdma, REG_TX_DMA_IDX(qid), TX_RING_DMA_IDX_MASK,
--- a/drivers/net/ethernet/airoha/airoha_eth.h
+++ b/drivers/net/ethernet/airoha/airoha_eth.h
@@ -202,6 +202,7 @@ struct airoha_queue {
 	spinlock_t lock;
 	struct airoha_queue_entry *entry;
 	struct airoha_qdma_desc *desc;
+	dma_addr_t dma_addr;
 	u16 head;
 	u16 tail;
 
