--- a/drivers/net/ethernet/airoha/airoha_eth.c
+++ b/drivers/net/ethernet/airoha/airoha_eth.c
@@ -1232,11 +1232,38 @@ unlock:
 	return done;
 }
 
+static void airoha_qdma_setup_tx_queue(struct airoha_queue *q)
+{
+	struct airoha_qdma *qdma = q->qdma;
+	int i, qid = q - &qdma->q_tx[0];
+	struct airoha_queue_entry *e;
+	int index;
+
+	for (i = 0; i < q->ndesc; i++) {
+		u32 val = FIELD_PREP(QDMA_DESC_DONE_MASK, 1);
+
+		WRITE_ONCE(q->desc[i].ctrl, cpu_to_le32(val));
+	}
+
+	/* xmit ring drop default setting */
+	airoha_qdma_set(qdma, REG_TX_RING_BLOCKING(qid),
+			TX_RING_IRQ_BLOCKING_TX_DROP_EN_MASK);
+
+	airoha_qdma_wr(qdma, REG_TX_RING_BASE(qid), q->dma_addr);
+
+	e = list_first_entry(&q->tx_list, struct airoha_queue_entry, list);
+	index = e - q->entry;
+	airoha_qdma_rmw(qdma, REG_TX_CPU_IDX(qid), TX_RING_CPU_IDX_MASK,
+			FIELD_PREP(TX_RING_CPU_IDX_MASK, index));
+	airoha_qdma_rmw(qdma, REG_TX_DMA_IDX(qid), TX_RING_DMA_IDX_MASK,
+			FIELD_PREP(TX_RING_DMA_IDX_MASK, index));
+}
+
 static int airoha_qdma_init_tx_queue(struct airoha_queue *q,
 				     struct airoha_qdma *qdma, int size)
 {
 	struct airoha_eth *eth = qdma->eth;
-	int i, qid = q - &qdma->q_tx[0];
+	int i;
 
 	spin_lock_init(&q->lock);
 	q->qdma = qdma;
@@ -1253,23 +1280,11 @@ static int airoha_qdma_init_tx_queue(str
 	if (!q->desc)
 		return -ENOMEM;
 
-	for (i = 0; i < size; i++) {
-		u32 val = FIELD_PREP(QDMA_DESC_DONE_MASK, 1);
-
+	for (i = 0; i < size; i++)
 		list_add_tail(&q->entry[i].list, &q->tx_list);
-		WRITE_ONCE(q->desc[i].ctrl, cpu_to_le32(val));
-	}
-	q->ndesc = size;
 
-	/* xmit ring drop default setting */
-	airoha_qdma_set(qdma, REG_TX_RING_BLOCKING(qid),
-			TX_RING_IRQ_BLOCKING_TX_DROP_EN_MASK);
-
-	airoha_qdma_wr(qdma, REG_TX_RING_BASE(qid), q->dma_addr);
-	airoha_qdma_rmw(qdma, REG_TX_CPU_IDX(qid), TX_RING_CPU_IDX_MASK,
-			FIELD_PREP(TX_RING_CPU_IDX_MASK, 0));
-	airoha_qdma_rmw(qdma, REG_TX_DMA_IDX(qid), TX_RING_DMA_IDX_MASK,
-			FIELD_PREP(TX_RING_DMA_IDX_MASK, 0));
+	q->ndesc = size;
+	airoha_qdma_setup_tx_queue(q);
 
 	return 0;
 }
