--- a/drivers/net/ethernet/airoha/airoha_eth.h
+++ b/drivers/net/ethernet/airoha/airoha_eth.h
@@ -47,7 +47,7 @@
 
 #define AIROHA_LRO_PAGE_ORDER		order_base_2(SZ_16K / PAGE_SIZE)
 #define AIROHA_MAX_NUM_LRO_QUEUES	8
-#define AIROHA_RXQ_LRO_EN_MASK		0xff000000
+#define AIROHA_RXQ_LRO_EN_MASK		0xff
 #define AIROHA_RXQ_LRO_MAX_AGG_COUNT	64
 #define AIROHA_RXQ_LRO_MAX_AGG_TIME	100
 #define AIROHA_RXQ_LRO_MAX_AGE_TIME	2000 /* 1ms */
