--- a/drivers/net/ethernet/airoha/airoha_eth.c
+++ b/drivers/net/ethernet/airoha/airoha_eth.c
@@ -765,6 +765,7 @@ static int airoha_qdma_lro_rx_process(st
 		}
 	}
 
+	//printk("[%s-%d] COUNT:%d\n", __func__, __LINE__, (int)FIELD_GET(QDMA_ETH_RXMSG_AGG_COUNT_MASK, msg2));
 	return 0;
 }
 
@@ -2475,6 +2476,7 @@ static int airoha_dev_set_features(struc
 	u32 val, pse_port, gdm_rxchn_en;
 	struct airoha_queue *q;
 
+	printk("[%s-%d]\n", __func__, __LINE__);
 	if (!(diff & NETIF_F_LRO))
 		return 0;
 
@@ -2499,12 +2501,14 @@ static int airoha_dev_set_features(struc
 	/* reset Tx queues */
 	for (i = 0; i < ARRAY_SIZE(qdma->q_tx); i++) {
 		netdev_tx_reset_subqueue(netdev, i);
-		if (!qdma->q_tx[i].ndesc)
-			continue;
+		//if (!qdma->q_tx[i].ndesc)
+		//	continue;
 
-		airoha_qdma_cleanup_tx_queue(&qdma->q_tx[i]);
+		//airoha_qdma_cleanup_tx_queue(&qdma->q_tx[i]);
 	}
 
+	printk("[%s-%d] XXX\n", __func__, __LINE__);
+
 	/* disable FE GDMA channel */
 	gdm_rxchn_en = airoha_fe_rr(eth, REG_GDM_RXCHN_EN(port->id));
 	airoha_fe_wr(eth, REG_GDM_RXCHN_EN(port->id), 0x0);
@@ -2521,6 +2525,7 @@ static int airoha_dev_set_features(struc
 	if (err)
 		return err;
 
+	printk("[%s-%d]\n", __func__, __LINE__);
 	/* disable Rx DMA */
 	airoha_qdma_clear(qdma, REG_QDMA_GLOBAL_CFG,
 			  GLOBAL_CFG_RX_DMA_EN_MASK);
@@ -2538,15 +2543,17 @@ static int airoha_dev_set_features(struc
 		airoha_qdma_rx_queue_deinit(q);
 	}
 
+	printk("[%s-%d]\n", __func__, __LINE__);
+
 	/* reset LRO configuration */
 	airoha_fe_lro_disable(eth, qdma_id);
 	/* reconfigure tx queues */
 	airoha_qdma_reset_tx(qdma);
 
+	printk("[%s-%d]\n", __func__, __LINE__);
+
 	/* reconfigure rx queues */
 	for (i = 0; i < ARRAY_SIZE(qdma->q_rx); i++) {
-		int err;
-
 		q = &qdma->q_rx[i];
 		if (!q->ndesc)
 			continue;
@@ -2560,6 +2567,7 @@ static int airoha_dev_set_features(struc
 
 		airoha_qdma_setup_rx_queue(q);
 		if (features & NETIF_F_LRO) {
+			printk("[%s-%d] Q(%d) %d\n", __func__, __LINE__, i, lro_queue_index);
 			airoha_fe_lro_init_rx_queue(eth, qdma_id,
 						    lro_queue_index, i,
 						    q->page_pool->p.pool_size,
@@ -2570,6 +2578,8 @@ static int airoha_dev_set_features(struc
 		napi_enable(&q->napi);
 	}
 
+	printk("[%s-%d]\n", __func__, __LINE__);
+
 	err = airoha_qdma_init_hfwd_queues(qdma);
 	if (err)
 		return err;
@@ -2587,6 +2597,8 @@ static int airoha_dev_set_features(struc
 	airoha_qdma_set(qdma, REG_QDMA_GLOBAL_CFG,
 			GLOBAL_CFG_TX_DMA_EN_MASK | GLOBAL_CFG_RX_DMA_EN_MASK);
 
+	printk("[%s-%d]\n", __func__, __LINE__);
+
 	return 0;
 }
 
