diff -uNr old/net/mac80211/iface.c new/net/mac80211/iface.c
--- old/net/mac80211/iface.c	2025-04-24 14:05:42.599131000 +0800
+++ new/net/mac80211/iface.c	2025-04-24 14:26:37.939030000 +0800
@@ -987,7 +987,9 @@
 	.ndo_start_xmit		= ieee80211_subif_start_xmit_8023,
 	.ndo_set_rx_mode	= ieee80211_set_multicast_list,
 	.ndo_set_mac_address	= ieee80211_change_mac,
+#ifdef CONFIG_NET_MEDIATEK_SOC_WED		
 	.ndo_fill_forward_path	= ieee80211_netdev_fill_forward_path,
+#endif	
 #if LINUX_VERSION_IS_LESS(6,6,0)
 	.ndo_fill_receive_path	= ieee80211_netdev_fill_receive_path,
 #endif
