--- a/drivers/net/ethernet/airoha/airoha_npu.c
+++ b/drivers/net/ethernet/airoha/airoha_npu.c
@@ -12,7 +12,6 @@
 #include <linux/of_reserved_mem.h>
 #include <linux/regmap.h>
 #include <linux/soc/airoha/airoha_offload.h>
-#include <net/page_pool/helpers.h>
 
 #include "airoha_eth.h"
 
@@ -665,109 +664,6 @@ static int airoha_npu_wlan_get_npu_suppo
 				       map);
 }
 
-static int airoha_npu_wlan_fill_rx_queue(struct airoha_npu *npu,
-					 struct airoha_npu_queue *q)
-{
-	enum dma_data_direction dir = page_pool_get_dma_dir(q->page_pool);
-	int nframes = 0;
-
-	while (q->queued < q->ndesc - 1) {
-		struct airoha_npu_rx_dma_desc *desc = &q->desc[q->head];
-		struct airoha_npu_queue_entry *e = &q->entry[q->head];
-		struct page *page;
-		int offset;
-
-		page = page_pool_dev_alloc_frag(q->page_pool, &offset,
-						q->buf_size);
-		if (!page)
-			return -ENOMEM;
-
-		e->buf = page_address(page) + offset;
-		e->dma_addr = page_pool_get_dma_addr(page) + offset;
-		e->dma_len = SKB_WITH_OVERHEAD(q->buf_size);
-
-		memset(desc, 0, sizeof(*desc));
-		desc->addr = e->dma_addr;
-
-		dma_sync_single_for_device(npu->dev, e->dma_addr, e->dma_len,
-					   dir);
-
-		q->head = (q->head + 1) % q->ndesc;
-		q->queued++;
-	}
-
-	return nframes;
-}
-
-static struct sk_buff *airoha_npu_wlan_rx_dequeue(struct airoha_npu *npu,
-						  int qid, u16 *hash,
-						  u16 *reason)
-{
-	struct airoha_npu_rx_dma_desc *desc;
-	enum dma_data_direction dir;
-	struct airoha_npu_queue *q;
-	struct sk_buff *skb = NULL;
-	int i, nframes, index;
-
-	if (qid >= ARRAY_SIZE(npu->q_rx))
-		return NULL;
-
-	q = &npu->q_rx[qid];
-
-	dir = page_pool_get_dma_dir(q->page_pool);
-	desc = &q->desc[q->tail];
-	index = q->tail;
-
-	nframes = FIELD_GET(NPU_RX_DMA_PKT_COUNT_MASK, desc->info);
-	nframes = max_t(int, nframes, 1);
-
-	for (i = 0; i < nframes; i++) {
-		struct airoha_npu_queue_entry *e = &q->entry[index];
-		struct page *page = virt_to_head_page(e->buf);
-		int len = FIELD_GET(NPU_RX_DMA_DESC_CUR_LEN_MASK,
-				    desc->ctrl);
-
-		if (!FIELD_GET(NPU_RX_DMA_DESC_DONE_MASK, desc->ctrl)) {
-			dev_kfree_skb(skb);
-			return NULL;
-		}
-
-		dma_sync_single_for_cpu(npu->dev, desc->addr,
-					SKB_WITH_OVERHEAD(q->buf_size), dir);
-
-		if (!skb) {
-			skb = build_skb(e->buf, q->buf_size);
-			if (!skb)
-				return NULL;
-
-			__skb_put(skb, len);
-			skb_mark_for_recycle(skb);
-			skb_reset_mac_header(skb);
-		} else {
-			struct skb_shared_info *shinfo = skb_shinfo(skb);
-			int nr_frags = shinfo->nr_frags;
-
-			if (nr_frags < ARRAY_SIZE(shinfo->frags))
-				skb_add_rx_frag(skb, nr_frags, page,
-						e->buf - page_address(page),
-						len, q->buf_size);
-		}
-
-		index = (index + 1) % q->ndesc;
-		desc = &q->desc[index];
-
-		*hash = FIELD_GET(NPU_RX_DMA_FOE_ID_MASK, desc->info);
-		*reason = FIELD_GET(NPU_RX_DMA_CRSN_MASK, desc->info);
-	}
-	q->tail = index;
-	q->queued -= i;
-
-	airoha_npu_wlan_fill_rx_queue(npu, q);
-	regmap_write(npu->regmap, REG_RX_DMA_IDX(qid), q->tail);
-
-	return skb;
-}
-
 static u32 airoha_npu_wlan_get_queue_addr(struct airoha_npu *npu, int qid,
 					  bool xmit)
 {
@@ -777,142 +673,32 @@ static u32 airoha_npu_wlan_get_queue_add
 	return REG_RX_BASE(qid);
 }
 
-static int airoha_npu_wlan_init_rx_queue(struct airoha_npu *npu,
-					 struct airoha_npu_queue *q)
-{
-	const struct page_pool_params pp_params = {
-		.order = 0,
-		.pool_size = 256,
-		.flags = PP_FLAG_DMA_MAP | PP_FLAG_DMA_SYNC_DEV |
-			 PP_FLAG_PAGE_FRAG,
-		.dma_dir = DMA_FROM_DEVICE,
-		.max_len = PAGE_SIZE,
-		.nid = NUMA_NO_NODE,
-		.dev = npu->dev,
-	};
-	struct airoha_npu_rx_dma_desc *rx_desc;
-	struct airoha_npu_queue_entry *entry;
-	int ndesc, qid = q - &npu->q_rx[0];
-	dma_addr_t dma_addr;
-
-	ndesc = qid == 1 ? NPU_RX1_DESC_NUM : NPU_RX0_DESC_NUM;
-	rx_desc = dmam_alloc_coherent(npu->dev, ndesc * sizeof(*rx_desc),
-				      &dma_addr, GFP_KERNEL);
-	if (!rx_desc)
-		return -ENOMEM;
-
-	entry = devm_kzalloc(npu->dev, ndesc * sizeof(*entry), GFP_KERNEL);
-	if (!entry)
-		return -ENOMEM;
-
-	q->page_pool = page_pool_create(&pp_params);
-	if (IS_ERR(q->page_pool)) {
-		int err = PTR_ERR(q->page_pool);
-
-		q->page_pool = NULL;
-		return err;
-	}
-
-	q->desc = rx_desc;
-	q->entry = entry;
-	q->ndesc = ndesc;
-	q->buf_size = PAGE_SIZE / 2;
-	airoha_npu_wlan_fill_rx_queue(npu, q);
-
-	regmap_write(npu->regmap, REG_RX_DSCP_NUM(qid), ndesc);
-	regmap_write(npu->regmap, REG_RX_DMA_IDX(qid), 0);
-	regmap_write(npu->regmap, REG_RX_CPU_IDX(qid), 0);
-	regmap_write(npu->regmap, REG_RX_BASE(qid), dma_addr);
-
-	return 0;
-}
-
-static void airoha_npu_cleanup_rx_queue(struct airoha_npu *npu,
-					struct airoha_npu_queue *q)
-{
-	enum dma_data_direction dir = page_pool_get_dma_dir(q->page_pool);
-
-	while (q->queued) {
-		struct airoha_npu_queue_entry *e = &q->entry[q->tail];
-
-		dma_sync_single_for_cpu(npu->dev, e->dma_addr, e->dma_len,
-					dir);
-		page_pool_put_full_page(q->page_pool,
-					virt_to_head_page(e->buf),
-					false);
-		q->tail = (q->tail + 1) % q->ndesc;
-		q->queued--;
-	}
-}
-
-static void airoha_npu_cleanup_rx_queues(struct airoha_npu *npu)
-{
-	int i;
-
-	for (i = 0; i < ARRAY_SIZE(npu->q_rx); i++) {
-		airoha_npu_cleanup_rx_queue(npu, &npu->q_rx[i]);
-		if (npu->q_rx[i].page_pool)
-			page_pool_destroy(npu->q_rx[i].page_pool);
-	}
-}
-
-static int airoha_npu_wlan_init_rx_queues(struct airoha_npu *npu)
-{
-	int i, err = 0;
-
-	for (i = 0; i < ARRAY_SIZE(npu->q_rx); i++) {
-		err = airoha_npu_wlan_init_rx_queue(npu, &npu->q_rx[i]);
-		if (err)
-			goto cleanup_rx_queues;
-	}
-
-	return 0;
-
-cleanup_rx_queues:
-	airoha_npu_cleanup_rx_queues(npu);
-
-	return err;
-}
-
 static int airoha_npu_wlan_init(struct airoha_npu *npu)
 {
 	int err;
 
-	err = airoha_npu_wlan_init_rx_queues(npu);
-	if (err)
-		return err;
-
 	err = airoha_npu_wlan_send_msg(npu, 1,
 				       WLAN_FUNC_SET_WAIT_NPU_BAND0_ONCPU, 0);
 	if (err)
-		goto cleanup_rx_queues;
+		return err;
 
 	err = airoha_npu_wlan_set_reserved_memory(npu, 0, "tx-bufid",
 			WLAN_FUNC_SET_WAIT_TX_BUF_CHECK_ADDR);
 	if (err)
-		goto cleanup_rx_queues;
+		return err;
 
 	err = airoha_npu_wlan_set_reserved_memory(npu, 0, "pkt",
 			WLAN_FUNC_SET_WAIT_PKT_BUF_ADDR);
 	if (err)
-		goto cleanup_rx_queues;
+		return err;
 
 	err = airoha_npu_wlan_set_reserved_memory(npu, 0, "tx-pkt",
 			WLAN_FUNC_SET_WAIT_TX_PKT_BUF_ADDR);
 	if (err)
-		goto cleanup_rx_queues;
-
-	err = airoha_npu_wlan_send_msg(npu, 0,
-				       WLAN_FUNC_SET_WAIT_IS_FORCE_TO_CPU, 0);
-	if (err)
-		goto cleanup_rx_queues;
-
-	return 0;
-
-cleanup_rx_queues:
-	airoha_npu_cleanup_rx_queues(npu);
+		return err;
 
-	return err;
+	return airoha_npu_wlan_send_msg(npu, 0,
+					WLAN_FUNC_SET_WAIT_IS_FORCE_TO_CPU, 0);
 }
 
 static void airoha_npu_wlan_set_irq_mask(struct airoha_npu *npu, int q)
@@ -1058,7 +844,6 @@ static int airoha_npu_probe(struct platf
 		airoha_npu_wlan_set_rx_ring_for_txdone;
 	npu->ops.wlan_get_npu_support_map =
 		airoha_npu_wlan_get_npu_support_map;
-	npu->ops.wlan_rx_dequeue = airoha_npu_wlan_rx_dequeue;
 	npu->ops.wlan_get_queue_addr = airoha_npu_wlan_get_queue_addr;
 	npu->ops.wlan_set_irq_mask = airoha_npu_wlan_set_irq_mask;
 	npu->ops.wlan_get_irq = airoha_npu_wlan_get_irq;
@@ -1155,8 +940,6 @@ static void airoha_npu_remove(struct pla
 
 	for (i = 0; i < ARRAY_SIZE(npu->cores); i++)
 		cancel_work_sync(&npu->cores[i].wdt_work);
-
-	airoha_npu_cleanup_rx_queues(npu);
 }
 
 static struct platform_driver airoha_npu_driver = {
--- a/include/linux/soc/airoha/airoha_offload.h
+++ b/include/linux/soc/airoha/airoha_offload.h
@@ -81,7 +81,6 @@ static inline void airoha_ppe_put(struct
 
 #define NPU_NUM_CORES		8
 #define NPU_NUM_IRQ		6
-#define NPU_NUM_RXQ		2
 #define NPU_RX0_DESC_NUM	512
 #define NPU_RX1_DESC_NUM	512
 
@@ -127,25 +126,6 @@ struct airoha_npu_tx_dma_desc {
 	u8 txwi[NPU_TXWI_LEN];
 } __packed;
 
-struct airoha_npu_queue_entry {
-	void *buf;
-	dma_addr_t dma_addr;
-	u16 dma_len;
-};
-
-struct airoha_npu_queue {
-	u16 head;
-	u16 tail;
-
-	int queued;
-	int ndesc;
-	int buf_size;
-
-	struct airoha_npu_queue_entry *entry;
-	struct airoha_npu_rx_dma_desc *desc;
-	struct page_pool *page_pool;
-};
-
 struct airoha_npu {
 #if (IS_BUILTIN(CONFIG_NET_AIROHA_NPU) || IS_MODULE(CONFIG_NET_AIROHA_NPU))
 	struct device *dev;
@@ -162,8 +142,6 @@ struct airoha_npu {
 
 	struct airoha_foe_stats *stats;
 
-	struct airoha_npu_queue q_rx[NPU_NUM_RXQ];
-
 	struct {
 		int (*ppe_init)(struct airoha_npu *npu);
 		int (*ppe_deinit)(struct airoha_npu *npu);
@@ -197,9 +175,6 @@ struct airoha_npu {
 						   int index, u32 addr);
 		int (*wlan_get_npu_support_map)(struct airoha_npu *npu,
 						int index, u32 *map);
-		struct sk_buff *(*wlan_rx_dequeue)(struct airoha_npu *npu,
-						   int index, u16 *hash,
-						   u16 *reason);
 		u32 (*wlan_get_queue_addr)(struct airoha_npu *npu, int qid,
 					   bool xmit);
 		void (*wlan_set_irq_mask)(struct airoha_npu *npu, int q);
--- a/drivers/net/ethernet/airoha/Kconfig
+++ b/drivers/net/ethernet/airoha/Kconfig
@@ -11,7 +11,6 @@ config NET_AIROHA_NPU
 	tristate "Airoha NPU support"
 	select WANT_DEV_COREDUMP
 	select REGMAP_MMIO
-	select PAGE_POOL
 	help
 	  This driver supports Airoha Network Processor (NPU) available
 	  on the Airoha Soc family.
