--- a/drivers/net/ethernet/airoha/airoha_npu.c
+++ b/drivers/net/ethernet/airoha/airoha_npu.c
@@ -493,7 +493,7 @@ static int airoha_npu_wlan_send_msg(stru
 	struct wlan_mbox_data *wlan_data;
 	int err;
 
-	wlan_data = kzalloc(sizeof(*wlan_data), GFP_KERNEL);
+	wlan_data = kzalloc(sizeof(*wlan_data), GFP_ATOMIC);
 	if (!wlan_data)
 		return -ENOMEM;
 
@@ -515,7 +515,7 @@ static int airoha_npu_wlan_get_msg(struc
 	struct wlan_mbox_data *wlan_data;
 	int err;
 
-	wlan_data = kzalloc(sizeof(*wlan_data), GFP_KERNEL);
+	wlan_data = kzalloc(sizeof(*wlan_data), GFP_ATOMIC);
 	if (!wlan_data)
 		return -ENOMEM;
 
@@ -578,7 +578,7 @@ static int airoha_npu_wlan_set_txrx_reg_
 	struct wlan_mbox_data *wlan_data;
 	int err;
 
-	wlan_data = kzalloc(sizeof(*wlan_data), GFP_KERNEL);
+	wlan_data = kzalloc(sizeof(*wlan_data), GFP_ATOMIC);
 	if (!wlan_data)
 		return -ENOMEM;
 
