--- a/drivers/net/ethernet/airoha/airoha_npu.c
+++ b/drivers/net/ethernet/airoha/airoha_npu.c
@@ -49,6 +49,11 @@
 #define REG_IRQ_RXDONE(_n)		(NPU_WLAN_BASE_ADDR + ((_n) << 2) + 0x034)
 #define NPU_IRQ_RX_MASK(_n)		((_n) == 1 ? BIT(17) : BIT(16))
 
+#define REG_TX_BASE(_n)			(NPU_WLAN_BASE_ADDR + ((_n) << 4) + 0x080)
+#define REG_TX_DSCP_NUM(_n)		(NPU_WLAN_BASE_ADDR + ((_n) << 4) + 0x084)
+#define REG_TX_DMA_IDX(_n)		(NPU_WLAN_BASE_ADDR + ((_n) << 4) + 0x088)
+#define REG_TX_CPU_IDX(_n)		(NPU_WLAN_BASE_ADDR + ((_n) << 4) + 0x08c)
+
 #define NPU_TIMER_BASE_ADDR		0x310100
 #define REG_WDT_TIMER_CTRL(_n)		(NPU_TIMER_BASE_ADDR + ((_n) * 0x100))
 #define WDT_EN_MASK			BIT(25)
@@ -213,6 +218,13 @@ struct wlan_mbox_data {
 	};
 };
 
+struct npu_tx_dma_desc {
+	__le32 ctrl;
+	dma_addr_t addr;
+	__le64 rsv;
+	u8 data[6];
+};
+
 static struct airoha_npu __rcu *npu_ptr;
 
 static int airoha_npu_send_msg(struct airoha_npu *npu, int func_id,
@@ -656,8 +668,31 @@ static int airoha_npu_wlan_get_npu_suppo
 
 static int airoha_npu_wlan_init(struct airoha_npu *npu)
 {
+	struct npu_tx_dma_desc *tx_desc;
+	dma_addr_t dma_addr;
 	int err;
 
+	tx_desc = dmam_alloc_coherent(npu->dev,
+				      NPU_TX_DESC_NUM * sizeof(*tx_desc),
+				      &dma_addr, GFP_KERNEL);
+	if (!tx_desc)
+		return -ENOMEM;
+
+	npu->tx_desc = tx_desc;
+
+	regmap_write(npu->regmap, REG_TX_DSCP_NUM(2), NPU_TX0_DESC_NUM);
+	regmap_write(npu->regmap, REG_TX_DSCP_NUM(3), NPU_TX1_DESC_NUM);
+
+	regmap_write(npu->regmap, REG_TX_CPU_IDX(2), 0);
+	regmap_write(npu->regmap, REG_TX_CPU_IDX(3), 0);
+
+	regmap_write(npu->regmap, REG_TX_DMA_IDX(2), 0);
+	regmap_write(npu->regmap, REG_TX_DMA_IDX(3), 0);
+
+	regmap_write(npu->regmap, REG_TX_BASE(2), dma_addr);
+	regmap_write(npu->regmap, REG_TX_BASE(3),
+		     dma_addr + NPU_TX0_DESC_NUM * sizeof(*tx_desc));
+
 	err = airoha_npu_wlan_send_msg(npu, 1,
 				       WLAN_FUNC_SET_WAIT_NPU_BAND0_ONCPU, 0);
 	if (err)
--- a/include/linux/soc/airoha/airoha_npu.h
+++ b/include/linux/soc/airoha/airoha_npu.h
@@ -37,6 +37,10 @@
 #define NPU_NUM_CORES		8
 #define NPU_NUM_IRQ		6
 
+#define NPU_TX0_DESC_NUM	1024
+#define NPU_TX1_DESC_NUM	1024
+#define NPU_TX_DESC_NUM		(NPU_TX0_DESC_NUM + NPU_TX1_DESC_NUM)
+
 struct airoha_npu {
 #if (IS_BUILTIN(CONFIG_NET_AIROHA_NPU) || IS_MODULE(CONFIG_NET_AIROHA_NPU))
 	struct device *dev;
@@ -53,6 +57,8 @@ struct airoha_npu {
 
 	struct airoha_foe_stats *stats;
 
+	void *tx_desc;
+
 	struct {
 		int (*ppe_init)(struct airoha_npu *npu);
 		int (*ppe_deinit)(struct airoha_npu *npu);
