--- a/drivers/net/ethernet/airoha/airoha_npu.c
+++ b/drivers/net/ethernet/airoha/airoha_npu.c
@@ -846,9 +846,6 @@ static int airoha_npu_probe(struct platf
 	npu->ops.hostadpt_rx_handler_ops = hostadpt_rx_handler;
 	npu->ops.hostdapt_registe_wifitask_ops = hostdapt_registe_wifitask;
 
-#ifdef CONFIG_NET_AIROHA_SOC_NPU
-	npu->Regbase = base;
-#endif
 	npu->regmap = devm_regmap_init_mmio(dev, base, &regmap_config);
 	if (IS_ERR(npu->regmap))
 		return PTR_ERR(npu->regmap);
--- a/drivers/net/ethernet/airoha/hostadpt.c
+++ b/drivers/net/ethernet/airoha/hostadpt.c
@@ -77,17 +77,6 @@ u32 hostadpt_rx_cpu_idx[HOSTADPT_RX_RING
 extern struct airoha_eth *glb_eth;
 static struct airoha_npu *npu;
 
-static u32 get_npu_hostadpt_reg(int hadap_ofs)
-{
-	return readl(npu->Regbase + NPU_HOSTADAPTER_BASE+hadap_ofs);
-}
-
-static u32 set_npu_hostadpt_reg(int hadap_ofs,int val)
-{
-	writel(val, npu->Regbase + NPU_HOSTADAPTER_BASE+hadap_ofs);
-	return 0;
-}
-
 #if defined(TCSUPPORT_WLAN_INODE) || defined(TCSUPPORT_NPU_WIFI_TX)
 u32 hostadpt_tx_dma_idx[2] = {0};
 u32 hostadpt_tx_cpu_idx[2] = {0};
@@ -161,20 +150,20 @@ static int hostadpt_rxdscp_init(void)
 	
 	
 	//write rx0/rx1 num to reg
-	set_npu_hostadpt_reg(HOSTADPT_RX0_MAX_CNT, HOSTADPT_RX0_DSCP_NUM);
-	set_npu_hostadpt_reg(HOSTADPT_RX1_MAX_CNT, HOSTADPT_RX1_DSCP_NUM);
+	regmap_write(npu->regmap, NPU_HOSTADAPTER_BASE + HOSTADPT_RX0_MAX_CNT, HOSTADPT_RX0_DSCP_NUM);
+	regmap_write(npu->regmap, NPU_HOSTADAPTER_BASE + HOSTADPT_RX1_MAX_CNT, HOSTADPT_RX1_DSCP_NUM);
 	
 	//set rx0/rx1 hostadpt_rx_cpu_idx/hostadpt_rx_dma_idx 
-	set_npu_hostadpt_reg(HOSTADPT_RX0_CPU_IDX, 0);
-	set_npu_hostadpt_reg(HOSTADPT_RX1_CPU_IDX, 0);
+	regmap_write(npu->regmap, NPU_HOSTADAPTER_BASE + HOSTADPT_RX0_CPU_IDX, 0);
+	regmap_write(npu->regmap, NPU_HOSTADAPTER_BASE + HOSTADPT_RX1_CPU_IDX, 0);
 	
-	set_npu_hostadpt_reg(HOSTADPT_RX0_DMA_IDX, 0);
-	set_npu_hostadpt_reg(HOSTADPT_RX1_DMA_IDX, 0);
+	regmap_write(npu->regmap, NPU_HOSTADAPTER_BASE + HOSTADPT_RX0_DMA_IDX, 0);
+	regmap_write(npu->regmap, NPU_HOSTADAPTER_BASE + HOSTADPT_RX1_DMA_IDX, 0);
 	
 	//write rx0/rx1 base to reg
 	//Please add other code before this line which is about initialization. 
-	set_npu_hostadpt_reg(HOSTADPT_RX0_BASE_PTR, dscpDmaAddr);
-	set_npu_hostadpt_reg(HOSTADPT_RX1_BASE_PTR, dscpDmaAddr + sizeof(HOSTADPT_DMA_DSCP_T)*HOSTADPT_RX0_DSCP_NUM);
+	regmap_write(npu->regmap, NPU_HOSTADAPTER_BASE + HOSTADPT_RX0_BASE_PTR, dscpDmaAddr);
+	regmap_write(npu->regmap, NPU_HOSTADAPTER_BASE + HOSTADPT_RX1_BASE_PTR, dscpDmaAddr + sizeof(HOSTADPT_DMA_DSCP_T)*HOSTADPT_RX0_DSCP_NUM);
 
 	return 0;
 
@@ -457,9 +446,9 @@ error:
 		}	
 	}
 	if(ring_idx == 0)
-		set_npu_hostadpt_reg(HOSTADPT_RX0_CPU_IDX, hostadpt_rx_cpu_idx[ring_idx]);	
+		regmap_write(npu->regmap, NPU_HOSTADAPTER_BASE + HOSTADPT_RX0_CPU_IDX, hostadpt_rx_cpu_idx[ring_idx]);	
 	else if(ring_idx == 1)
-		set_npu_hostadpt_reg(HOSTADPT_RX1_CPU_IDX, hostadpt_rx_cpu_idx[ring_idx]);	
+		regmap_write(npu->regmap, NPU_HOSTADAPTER_BASE + HOSTADPT_RX1_CPU_IDX, hostadpt_rx_cpu_idx[ring_idx]);	
 		
 	if(skb != NULL){	
 		skb_reset_mac_header(skb);
@@ -643,7 +632,7 @@ int hostadpt_tx_handler(unsigned char ba
 		//disable isr
 		spin_lock_irqsave(&hostadpt_lock_2G, flags);
 		HOSTADPT_MSG(DBG_LOG, "Send data to RX0 ... \n");
-		cpuIdx0 = get_npu_hostadpt_reg(HOSTADPT_TX2_CPU_IDX);//this will update by isr in npu
+		regmap_read(npu->regmap, NPU_HOSTADAPTER_BASE  + HOSTADPT_TX2_CPU_IDX, &cpuIdx0);
 		HOSTADPT_MSG(DBG_LOG, "HOSTADPT_TX2_CPU_IDX = %d\n", dmaIdx0);
 		pHostadptTX0 = pHostadptTX_BASE[0] + hostadpt_tx_dma_idx[0];
 		if(pHostadptTX0->pkt_ctrl.done == 0 && hostadpt_is_ring_available(hostadpt_tx_dma_idx[0],
@@ -655,7 +644,7 @@ int hostadpt_tx_handler(unsigned char ba
 			hostadpt_tx_dma_idx[0] = (hostadpt_tx_dma_idx[0] + 1);
 			if(hostadpt_tx_dma_idx[0] >= HOSTADPT_TX0_DSCP_NUM)
 				hostadpt_tx_dma_idx[0] = 0;
-			set_npu_hostadpt_reg(HOSTADPT_TX2_DMA_IDX, hostadpt_tx_dma_idx[0]);
+			regmap_write(npu->regmap, NPU_HOSTADAPTER_BASE + HOSTADPT_TX2_DMA_IDX, hostadpt_tx_dma_idx[0]);
 		}
 		else
 		{
@@ -687,7 +676,7 @@ int hostadpt_tx_handler(unsigned char ba
 	{
 		spin_lock_irqsave(&hostadpt_lock_test, flags);
 		HOSTADPT_MSG(DBG_LOG, "Send data to RX0 ... \n");
-		cpuIdx1 = get_npu_hostadpt_reg(HOSTADPT_TX3_CPU_IDX);
+		regmap_read(npu->regmap, NPU_HOSTADAPTER_BASE  + HOSTADPT_TX3_CPU_IDX, &cpuIdx1);
 		HOSTADPT_MSG(DBG_LOG, "HOSTADPT_TX3_DMA_IDX = %d\n", dmaIdx1);
 		pHostadptTX1 = pHostadptTX_BASE[1] + hostadpt_tx_dma_idx[1];
 		if(pHostadptTX1->pkt_ctrl.done == 0 && hostadpt_is_ring_available(hostadpt_tx_dma_idx[1],
@@ -699,7 +688,7 @@ int hostadpt_tx_handler(unsigned char ba
 			hostadpt_tx_dma_idx[1] = (hostadpt_tx_dma_idx[1] + 1);
 			if(hostadpt_tx_dma_idx[1] >= HOSTADPT_TX1_DSCP_NUM)
 				hostadpt_tx_dma_idx[1] = 0;
-			set_npu_hostadpt_reg(HOSTADPT_TX3_DMA_IDX, hostadpt_tx_dma_idx[1]);
+			regmap_write(npu->regmap, NPU_HOSTADAPTER_BASE + HOSTADPT_TX3_DMA_IDX, hostadpt_tx_dma_idx[1]);
 		}
 		else 
 		{
--- a/include/linux/soc/airoha/airoha_npu.h
+++ b/include/linux/soc/airoha/airoha_npu.h
@@ -41,7 +41,6 @@ struct airoha_npu {
 #if (IS_BUILTIN(CONFIG_NET_AIROHA_NPU) || IS_MODULE(CONFIG_NET_AIROHA_NPU))
 	struct device *dev;
 	struct regmap *regmap;
-	void __iomem *Regbase;
 
 	struct airoha_npu_core {
 		struct airoha_npu *npu;
