From 220a1aaad092faa2f9d87fe65532e9fda83f4d0a Mon Sep 17 00:00:00 2001
Message-ID: <220a1aaad092faa2f9d87fe65532e9fda83f4d0a.1741037135.git.lorenzo@kernel.org>
From: Lorenzo Bianconi <lorenzo@kernel.org>
Date: Mon, 3 Mar 2025 22:25:10 +0100
Subject: [PATCH] Revert "page_pool: remove PP_FLAG_PAGE_FRAG"

This reverts commit 3e85822b9c66490bd74e82e091249b659e7f52b1.
---
 mac80211.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mac80211.c b/mac80211.c
index 983b41bc..20afd9f0 100644
--- a/mac80211.c
+++ b/mac80211.c
@@ -613,7 +613,7 @@ int mt76_create_page_pool(struct mt76_dev *dev, struct mt76_queue *q)
 {
 	struct page_pool_params pp_params = {
 		.order = 0,
-		.flags = 0,
+		.flags = PP_FLAG_PAGE_FRAG,
 		.nid = NUMA_NO_NODE,
 		.dev = dev->dma_dev,
 	};
-- 
2.48.1

