--- a/agg-rx.c
+++ b/agg-rx.c
@@ -172,7 +172,11 @@ void mt76_rx_aggr_reorder(struct sk_buff
 	ackp = status->qos_ctl & IEEE80211_QOS_CTL_ACK_POLICY_MASK;
 	if (ackp == IEEE80211_QOS_CTL_ACK_POLICY_NOACK)
 		return;
-
+	
+	#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD
+		return; 
+	#endif 
+	
 	tid = rcu_dereference(wcid->aggr[tidno]);
 	if (!tid)
 		return;
--- a/dma.c
+++ b/dma.c
@@ -6,6 +6,9 @@
 #include <linux/dma-mapping.h>
 #include "mt76.h"
 #include "dma.h"
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD
+#include "mt76_npu_offload.h"
+#endif
 
 #if IS_ENABLED(CONFIG_NET_MEDIATEK_SOC_WED)
 
@@ -386,8 +389,15 @@ mt76_dma_tx_cleanup_idx(struct mt76_dev
 static void
 mt76_dma_kick_queue(struct mt76_dev *dev, struct mt76_queue *q)
 {
-	wmb();
-	Q_WRITE(q, cpu_idx, q->head);
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD
+#ifdef TCSUPPORT_NPU_WIFI_TX	
+	if(!(dev->phy.q_tx[0] == q ||(dev->phys[1] && dev->phys[1]->q_tx[0] == q)))
+#endif
+#endif
+	{
+		wmb();
+		Q_WRITE(q, cpu_idx, q->head);
+	}
 }
 
 static void
@@ -626,8 +636,13 @@ mt76_dma_tx_queue_skb(struct mt76_phy *p
 	if (ret < 0)
 		goto unmap;
 
+#ifdef TCSUPPORT_NPU_WIFI_TX
+	npu_mt76_dma_tx_queue_skb(phy, dev, t, skb, txwi, wcid, sta);
+	return -1;
+#else	
 	return mt76_dma_add_buf(dev, q, tx_info.buf, tx_info.nbuf,
 				tx_info.info, tx_info.skb, t);
+#endif
 
 unmap:
 	for (n--; n > 0; n--)
@@ -962,17 +977,55 @@ int mt76_dma_rx_poll(struct napi_struct
 {
 	struct mt76_dev *dev;
 	int qid, done = 0, cur;
-
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD	
+	void * pkt=NULL;
+	unsigned char bReschedule=0;
+	struct sk_buff *skb;
+	int npu_done=0;
+	unsigned int vend_spec = 0;
+//	union _2_hostapdt_vend_specific *vnd_spe;
+#endif
 	dev = container_of(napi->dev, struct mt76_dev, napi_dev);
 	qid = napi - dev->napi;
 
 	rcu_read_lock();
-
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD
+	if(qid != MT_RXQ_RRO_IND)
+	{
+		do {
+			cur = mt76_dma_rx_process(dev, &dev->q_rx[qid], budget - done);
+			mt76_rx_poll_complete(dev, qid, napi);
+			done += cur;
+		} while (cur && done < budget);
+	}
+	else{
+		while (npu_done < budget)
+		{		
+			if(fromHostadptPktHandle_hook)
+				pkt = fromHostadptPktHandle_hook(0, &bReschedule, &vend_spec);
+		
+			if(pkt)
+			{	npu_done++;
+				skb=(struct sk_buff *)pkt;
+				dev->drv->rx_skb(dev, &dev->q_rx[qid] - dev->q_rx, skb, 0);
+			}
+			else
+				break;
+		
+			mt76_rx_poll_complete(dev, qid, napi);	
+		}	
+		rcu_read_unlock();
+		if(npu_done < budget && napi_complete(napi))
+			dev->drv->rx_poll_complete(dev, qid);
+		return npu_done;
+	}
+#else
 	do {
 		cur = mt76_dma_rx_process(dev, &dev->q_rx[qid], budget - done);
 		mt76_rx_poll_complete(dev, qid, napi);
 		done += cur;
-	} while (cur && done < budget);
+	} while (cur && done < budget);	
+#endif
 
 	rcu_read_unlock();
 
--- a/Makefile
+++ b/Makefile
@@ -11,7 +11,7 @@ obj-$(CONFIG_MT792x_USB) += mt792x-usb.o
 
 mt76-y := \
 	mmio.o util.o trace.o dma.o mac80211.o debugfs.o eeprom.o \
-	tx.o agg-rx.o mcu.o wed.o scan.o channel.o
+	tx.o agg-rx.o mcu.o wed.o scan.o channel.o mt76_npu_offload.o
 
 mt76-$(CONFIG_PCI) += pci.o
 mt76-$(CONFIG_NL80211_TESTMODE) += testmode.o
--- a/mt76.h
+++ b/mt76.h
@@ -557,6 +557,11 @@ struct mt76_driver_ops {
 				struct ieee80211_vif *vif,
 				struct ieee80211_bss_conf *link_conf,
 				struct mt76_vif_link *mlink);
+#ifdef TCSUPPORT_NPU_WIFI_TX
+	void (*npu_txwi_free)(struct mt76_dev *dev, struct mt76_txwi_cache *t,
+			 struct ieee80211_sta *sta, struct mt76_wcid *wcid,
+			 struct list_head *free_list);
+#endif
 };
 
 struct mt76_channel_state {
@@ -984,6 +989,10 @@ struct mt76_dev {
 		struct mt76_usb usb;
 		struct mt76_sdio sdio;
 	};
+
+#if defined(TCSUPPORT_NPU_WIFI_OFFLOAD)
+	u32 slot_id;
+#endif		
 };
 
 /* per-phy stats.  */
--- /dev/null
+++ b/mt76_npu_offload.c
@@ -0,0 +1,426 @@
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD
+#include "mt76_npu_offload.h"
+
+unsigned long int  *dramReorderNodeBase;
+int setonce=0;
+
+
+int npu_enable_chk = 1;
+void npu_init_rxD(struct mt7996_dev *dev,struct pci_dev *pdev)
+{
+	/*
+
+	unsigned long int pcie_pa;	
+	int isMailBoxSuccess=0;
+	
+	u16 i;
+	
+	for (i = 0; i < ARRAY_SIZE(dev->wed_rro.addr_elem); i++)
+	{
+	
+		if(i < 3 || i > 125){
+			printk("=======npu_init_rxD========wed_rro.addr_elem[%d].phy_addr=%llx\n",
+				i, dev->wed_rro.addr_elem[i].phy_addr);
+		}
+		
+		isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_INODE_TXRX_REG_ADDR(0,
+			i, dev->wed_rro.addr_elem[i].phy_addr, 0, 0);
+		if(isMailBoxSuccess != 1){
+			printk("Error: WIFI_MAIL_API_SET_WAIT_INODE_TXRX_REG_ADDR 0 fail\n");
+		}
+		else {
+			continue;
+		}
+		
+	}
+	
+	isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_INODE_TXRX_REG_ADDR(1,
+		0, dev->wed_rro.session.phy_addr, 0, 0);
+	if(isMailBoxSuccess != 1){
+		printk("Error: WIFI_MAIL_API_SET_WAIT_INODE_TXRX_REG_ADDR 1 fail\n");
+	}
+	
+	pcie_pa = pci_resource_start(pdev, 0); //get pcie phy addr
+
+	isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_TX_RING_PCIE_ADDR(2, (pcie_pa + MT_RRO_ACK_SN_CTRL));
+*/
+
+}
+EXPORT_SYMBOL_GPL(npu_init_rxD);
+void wifi_offload_init_npu(struct pci_dev *pdev,struct mt7996_dev *dev,int flag)
+{
+
+	int isMailBoxSuccess = 0;
+	unsigned char band_idx = 0;	
+	unsigned long int pcie_pa;
+	dma_addr_t reorder_node_pa;
+	struct mt76_dev *mdev;
+	mdev = &dev->mt76;
+	struct mt7996_phy *phy;
+	phy=&dev->phy;
+	int npuOffloadEn=1;
+	u32 hif1_ofs = 0;
+
+	if(npuOffloadEn){
+		printk("npuOffload on mt76 init \n");	
+		if (pdev->bus){
+			mdev->slot_id = pci_domain_nr(pdev->bus);
+			printk("the value of the slot_id is %d\n", mdev->slot_id);
+		}
+		
+		if(mdev->slot_id)
+			isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_PCIE_PORT_TYPE(band_idx, P0__P1);
+		else
+			isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_PCIE_PORT_TYPE(band_idx, P0__P0);
+
+		if(isMailBoxSuccess != 1) 
+			printk("Error: KITE NPU_WIFI_OFFLOAD Set PCIE PORT Fail\n");	
+
+
+		/****************set global setting(only once)**************/
+		if((setonce != 1) && (setonce !=2))
+			setonce = 1;
+
+		if((setonce == 1)){
+			printk("!!!! do phy and virtual address mapping\n");
+
+			/**************ba node******************/
+			reorder_node_pa = npu_wifi_offload_get_ba_node_addr();
+			isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_DRAM_BA_NODE_ADDR(band_idx, reorder_node_pa);
+			if(isMailBoxSuccess != 1)
+				printk("Error: NPU_WIFI_OFFLOAD set BA Node in Dram base fail\n");
+			setonce = 2;
+		}
+		
+		pcie_pa = pci_resource_start(pdev, 0); //get pcie phy addr
+		if(is_mt7992(mdev))
+		{
+			printk("%s:%d mt7992 device !!!\n",__func__,__LINE__);
+			if (dev->hif2)
+				hif1_ofs = MT_WFDMA0_PCIE1(0) - MT_WFDMA0(0);
+#if	1		
+			/* npu rx rro ring*/
+			for(band_idx = 0; band_idx < 2; band_idx++){
+				if(band_idx){
+					isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_PCIE_ADDR(band_idx, (pcie_pa + MT_RXQ_RING_BASE(MT_RXQ_RRO_BAND1) + 0x90 + hif1_ofs));
+					
+					isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_DESC(band_idx, MT7996_RX_RING_SIZE);
+				}
+				else{
+					isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_PCIE_ADDR(band_idx, (pcie_pa + MT_RXQ_RING_BASE(MT_RXQ_RRO_BAND0) + 0x80));
+					isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_DESC(band_idx, MT7996_RX_RING_SIZE);
+				}
+			}
+			
+			band_idx = 0;
+
+			/* MSDU PG*/
+			isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_PCIE_ADDR(band_idx + 5, (pcie_pa + MT_RXQ_RING_BASE(MT_RXQ_MSDU_PAGE_BAND0) + 0xa0));
+			isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_DESC(band_idx + 5, MT7996_RX_RING_SIZE);
+
+			/* ind_cmd ring*/
+			isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_PCIE_ADDR(band_idx + 8, (pcie_pa + MT_RXQ_RRO_IND_RING_BASE));			
+			isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_DESC(band_idx + 8, MT7996_RX_RING_SIZE);
+			
+			isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_TX_RING_PCIE_ADDR(2, (pcie_pa + MT_RRO_ACK_SN_CTRL));
+#endif			
+			/* npu tx */
+#ifdef TCSUPPORT_NPU_WIFI_TX
+			for(band_idx = 0; band_idx < 2; band_idx++){
+				if(band_idx)
+					//isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_TX_RING_PCIE_ADDR(band_idx, (pcie_pa + MT_TXQ_RING_BASE(0) + 0x130 + hif1_ofs));
+					isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_TX_RING_PCIE_ADDR(band_idx, (pcie_pa + MT_TXQ_RING_BASE(0) + 0x150 + hif1_ofs));
+				else
+					isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_TX_RING_PCIE_ADDR(band_idx, (pcie_pa + MT_TXQ_RING_BASE(0) + 0x120));
+			}
+#endif			
+		}
+	}
+}
+EXPORT_SYMBOL_GPL(wifi_offload_init_npu);
+void get_npu_rxd(struct mt7996_dev *dev,struct pci_dev *pdev)
+{	
+	npu_info_t get_RingBase;
+	int isMailBoxSuccess=0;
+	unsigned char band_idx = 0;	
+	dma_addr_t npu_alloc_pa;
+	struct mt76_dev *mdev = &dev->mt76;
+	band_idx = mdev->slot_id;
+
+	/* npu rx rro ring*/
+	isMailBoxSuccess = WIFI_MAIL_API_GET_WAIT_RXDESC_BASE(0, &get_RingBase);
+	if(isMailBoxSuccess != 1) 
+		printk("Error: NPU_WIFI_OFFLOAD get ring0 base fail\n");
+	npu_alloc_pa=(NDIS_PHYSICAL_ADDRESS)get_RingBase.info;
+	writel(npu_alloc_pa,&dev->mt76.q_rx[MT_RXQ_RRO_BAND0].regs->desc_base);
+	
+	isMailBoxSuccess = WIFI_MAIL_API_GET_WAIT_RXDESC_BASE(1, &get_RingBase);
+	if(isMailBoxSuccess != 1) 
+		printk("Error: NPU_WIFI_OFFLOAD get ring1 base fail\n");
+	npu_alloc_pa=(NDIS_PHYSICAL_ADDRESS)get_RingBase.info;
+	writel(npu_alloc_pa,&dev->mt76.q_rx[MT_RXQ_RRO_BAND1].regs->desc_base);		
+
+	/* MSDU PG ring */
+	isMailBoxSuccess = WIFI_MAIL_API_GET_WAIT_RXDESC_BASE(10, &get_RingBase);
+	if(isMailBoxSuccess != 1) 
+		printk("Error: NPU_WIFI_OFFLOAD get MSDU PG base fail\n");
+	npu_alloc_pa=(NDIS_PHYSICAL_ADDRESS)get_RingBase.info;
+	writel(npu_alloc_pa,&dev->mt76.q_rx[MT_RXQ_MSDU_PAGE_BAND0].regs->desc_base);		
+	
+	/* ind_cmd ring */
+	isMailBoxSuccess = WIFI_MAIL_API_GET_WAIT_RXDESC_BASE(8, &get_RingBase);
+	if(isMailBoxSuccess != 1) 
+		printk("Error: NPU_WIFI_OFFLOAD get ind_cmd ring base fail\n");
+	npu_alloc_pa=(NDIS_PHYSICAL_ADDRESS)get_RingBase.info;
+	writel(npu_alloc_pa,&dev->mt76.q_rx[MT_RXQ_RRO_IND].regs->desc_base);	
+	
+}
+EXPORT_SYMBOL_GPL(get_npu_rxd);
+
+void npu_mt76_init_done(struct mt7996_dev *dev,struct pci_dev *pdev)
+{
+	int isMailBoxSuccess=0;
+	
+	/* rro set done */
+	isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_INODE_TXRX_REG_ADDR(2, 0, 0, 0, 0);
+	if (isMailBoxSuccess != 1)
+		printk("Error: WIFI_MAIL_API_SET_WAIT_INODE_TXRX_REG_ADDR 2 fail\n");
+	
+	isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_INODE_TXRX_REG_ADDR(7, 0, 0, 0, 0);
+	if (isMailBoxSuccess != 1)
+		printk("Error: WIFI_MAIL_API_SET_WAIT_INODE_TXRX_REG_ADDR 7 fail\n");
+
+}
+EXPORT_SYMBOL_GPL(npu_mt76_init_done);
+
+void npu_rro_reset_sessions(struct mt7996_dev *dev, u16 seid)
+{
+	int isMailBoxSuccess=0;
+
+	/* init_rro_addr_elem_by_seid */
+	isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_INODE_TXRX_REG_ADDR(3, seid, 0, 0, 0);
+	if(isMailBoxSuccess != 1){
+		printk("Error: WIFI_MAIL_API_SET_WAIT_INODE_TXRX_REG_ADDR 3 fail\n");
+	}
+}
+EXPORT_SYMBOL_GPL(npu_rro_reset_sessions);
+
+void get_npu_txd(struct mt7996_dev *dev,struct pci_dev *pdev)
+{
+	npu_info_t get_RingBase;
+	int isMailBoxSuccess=0;
+	unsigned char band_idx = 0;
+	struct mt76_dev *mdev = &dev->mt76;
+	
+	void *npu_alloc_va = NULL;
+	dma_addr_t npu_alloc_pa;
+	
+	void *txbuf_alloc_va_sp = NULL;
+	dma_addr_t txbuf_alloc_pa_sp;	
+	void *txbuf_alloc_va_fp = NULL;
+	dma_addr_t txbuf_alloc_pa_fp;
+	void *txbuf_alloc_va_swq = NULL;
+	dma_addr_t txbuf_alloc_pa_swq;
+	
+	for(band_idx = 0; band_idx < 2; band_idx++){
+		if(band_idx == 0){
+			isMailBoxSuccess = WIFI_MAIL_API_GET_WAIT_RXDESC_BASE((band_idx + 5), &get_RingBase);
+			npu_alloc_pa = (dma_addr_t)get_RingBase.info;
+			npu_alloc_va = ioremap((phys_addr_t)get_RingBase.info,4);					
+			writel(npu_alloc_pa, &dev->mt76.phy.q_tx[0]->regs->desc_base);
+		}
+		else{
+			isMailBoxSuccess = WIFI_MAIL_API_GET_WAIT_RXDESC_BASE((band_idx + 5), &get_RingBase);
+			npu_alloc_pa = (dma_addr_t)get_RingBase.info;
+			npu_alloc_va = ioremap((phys_addr_t)get_RingBase.info,4);					
+			writel(npu_alloc_pa, &dev->mt76.phys[1]->q_tx[0]->regs->desc_base);
+		}
+
+		txbuf_alloc_va_sp = dmam_alloc_coherent(mdev->dma_dev, MT7996_TX_RING_SIZE * 256,
+				      &txbuf_alloc_pa_sp, GFP_KERNEL);
+		if (!txbuf_alloc_va_sp) {
+			printk("%s can't alloc memory !\n", __func__);
+		}
+
+		txbuf_alloc_va_fp = dmam_alloc_coherent(mdev->dma_dev, MT7996_TX_RING_SIZE * 256,
+				      &txbuf_alloc_pa_fp, GFP_KERNEL);
+		if (!txbuf_alloc_va_fp) {
+			printk("%s can't alloc memory !\n", __func__);
+		}
+
+		//1024 is TX_PKT_INFO_MAX_NUM in npu_specific_define_eagle.h
+		txbuf_alloc_va_swq = dmam_alloc_coherent(mdev->dma_dev, 1024 * 256,
+				      &txbuf_alloc_pa_swq, GFP_KERNEL);
+		if (!txbuf_alloc_va_swq) {
+			printk("%s can't alloc memory !\n", __func__);
+		}
+		
+		isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_TX_BUF_SPACE_BASE(band_idx, txbuf_alloc_pa_sp);
+		if(isMailBoxSuccess != 1){
+			printk("Error: WIFI_MAIL_API_SET_WAIT_TX_BUF_SPACE_BASE fail\n");
+		}
+		
+		isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_TX_BUF_SPACE_BASE((band_idx+5), txbuf_alloc_pa_fp);
+		if(isMailBoxSuccess != 1){
+			printk("Error: WIFI_MAIL_API_SET_WAIT_TX_BUF_SPACE_BASE fail\n");
+		}
+		
+		isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_TX_BUF_SPACE_BASE((band_idx+10), txbuf_alloc_pa_swq);
+		if(isMailBoxSuccess != 1){
+			printk("Error: WIFI_MAIL_API_SET_WAIT_TX_BUF_SPACE_BASE fail\n");
+		}
+	}
+	
+}
+EXPORT_SYMBOL_GPL(get_npu_txd);
+
+char npu_mt76_token_release(struct mt76_dev *dev, unsigned int token, struct mt76_wcid *wcid, struct ieee80211_sta *sta)
+{	
+	struct mt76_txwi_cache *txwi_t;	
+	struct sk_buff *skb, *tmp;
+	LIST_HEAD(free_list);
+	
+	spin_lock_bh(&dev->token_lock);
+	
+	txwi_t = idr_remove(&dev->token, token);
+	
+	if (txwi_t) {
+		struct mt76_phy *phy = mt76_dev_phy(dev, txwi_t->phy_idx);
+		
+		dev->token_count--;
+		phy->tokens--;
+	}
+	
+	spin_unlock_bh(&dev->token_lock);
+	if (txwi_t) {
+		dev->drv->npu_txwi_free(dev, txwi_t, sta, wcid, &free_list);		
+		txwi_t->jiffies = 0;
+	}
+	
+	mt76_worker_schedule(&dev->tx_worker);
+	
+	list_for_each_entry_safe(skb, tmp, &free_list, list) {
+		skb_list_del_init(skb);
+		napi_consume_skb(skb, 1);
+	}
+
+	return 0;
+}
+
+void npu_mt76_dma_tx_queue_skb(struct mt76_phy *phy, struct mt76_dev *dev, struct mt76_txwi_cache *t, 
+	struct sk_buff *skb, void *txwi_ptr, struct mt76_wcid *wcid, struct ieee80211_sta *sta)
+{
+	struct sk_buff *n_skb = NULL;
+	unsigned short vnd_len;
+	__le32 *_txd;
+	struct port_info wifi_info;
+	unsigned char hostadpt_ring_num = 0, band_idx, bss_idx, is_tcp, qos_tid;
+	unsigned int wcidx, token;
+	npu_hif_v2_info npu_txp_info;
+	int ret = 0;//0 means to free skb
+	unsigned char txp_temp[MT_TXD_SIZE+HIF_TXP_SIZE];
+	_txd = (__le32 *)txwi_ptr;
+
+	n_skb = skb_clone(skb, GFP_ATOMIC);	
+
+	if (!n_skb) {
+		dev_kfree_skb(skb);
+		return;
+	}
+	
+	if(phy == dev->phys[1])		
+		band_idx = MT_BAND1;
+	else
+		band_idx = MT_BAND0;
+	
+	hostadpt_ring_num = band_idx;
+	token = (_txd[8] >> 16) & 0xffff;
+
+	if (ra_sw_nat_hook_tx != NULL)
+	{
+		memset(&wifi_info, 0 , sizeof(struct port_info));
+#if defined(TCSUPPORT_CPU_EN7580)
+		wifi_info.tsid = 0x7F;
+		wifi_info.nbq = 1;
+#endif
+		bss_idx = _txd[9] & 0xff; //txp[9] bss_id[7:0]
+		wcidx = (_txd[9] >> 8) & 0xffff; //txp[9] wtbl_idx[23:8]
+		//is_tcp = GET_PACKET_TCP(skb) ? 1 : 0;
+			
+		if (bss_idx > 0x7f) {
+			  printk("npu_mt76_dma_tx_queue_skb bss index is error\n");
+		}
+
+		qos_tid = n_skb->priority & IEEE80211_QOS_CTL_TID_MASK;
+
+		npu_txp_info._detail.ring_num = hostadpt_ring_num;
+		npu_txp_info._detail.qos_tid	=	qos_tid;
+
+		wifi_info.magic = FOE_MAGIC_WLAN;
+		wifi_info.udf = (bss_idx & 0x7f) | (is_tcp << 7);		
+		wifi_info.stag = ((wcidx & 0x7ff) | ((npu_txp_info.ctrl_word & 0x1f) << 11));
+
+#ifdef TCSUPPORT_MT7510_FE
+		if (ra_sw_nat_hook_tx(n_skb, &wifi_info, wifi_info.magic) == 0)
+#else
+		if (ra_sw_nat_hook_tx(n_skb, 1) == 0)
+#endif
+		{
+			if(npu_mt76_token_release(dev, token, wcid, sta)){
+				printk("npu_mt76_token_release fail\n");
+			}
+			dev_kfree_skb(n_skb);
+			return;
+		}
+	}
+
+	vnd_len = MT_TXD_SIZE + HIF_TXP_SIZE;
+	
+	memcpy(txp_temp,txwi_ptr,vnd_len);
+		
+	if(toHostadptPktHandle_hook && n_skb){
+		ret = toHostadptPktHandle_hook(hostadpt_ring_num, n_skb, txp_temp, vnd_len);
+	}
+	
+	//handle token/txwi free
+	if(npu_mt76_token_release(dev, token, wcid, sta)){
+		printk("npu_mt76_token_release fail\n");
+	}
+	
+	return;
+}
+
+EXPORT_SYMBOL_GPL(npu_mt76_dma_tx_queue_skb);
+
+void npu_set_rx_event(struct mt7996_dev *dev,struct pci_dev *pdev)
+{
+	unsigned long int pcie_pa = 0;
+	int isMailBoxSuccess = 0;
+	unsigned char band_idx = 0;
+	
+	isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_RX_RING_FOR_TXDONE_HW_BASE(band_idx, dev->mt76.q_rx[MT_RXQ_MAIN_WA].desc_dma); // npu  : tx_done_ring0_addr_uncache
+	
+	isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_DESC((band_idx + 10), MT7996_RX_MCU_RING_SIZE); //npu : g_tx_done_ring0_size
+	if(isMailBoxSuccess != 1){
+		printk( "%s() Error: WIFI_MAIL_API_SET_WAIT_DESC fail\n", __func__);
+	}
+	
+	pcie_pa = pci_resource_start(pdev, 0); //get pcie phy addr
+	isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_PCIE_ADDR((band_idx + 10), (pcie_pa + MT_RXQ_RING_BASE(MT_RXQ_MAIN_WA) + 0x20));//npu : TxDone_ring0_pcie_base
+	
+	return;
+}
+EXPORT_SYMBOL_GPL(npu_set_rx_event);
+
+int npu_enable(void)
+{
+	return npu_enable_chk;
+}
+void set_npu_enable(int enable)
+{
+	npu_enable_chk = enable;
+}
+EXPORT_SYMBOL_GPL(npu_enable);
+EXPORT_SYMBOL_GPL(set_npu_enable);
+
+#endif
--- /dev/null
+++ b/mt76_npu_offload.h
@@ -0,0 +1,88 @@
+#ifndef NPU_OFFLOAD_H
+#define NPU_OFFLOAD_H
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD	
+
+#include "modules/npu/wifi_mail.h"
+#include "mt7996/mt7996.h"
+#include "dma.h"
+#include "mt7996/mac.h"
+#include "mt76.h"
+
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/pci.h>
+
+
+#define BAND5G_IDX			1
+#define BAND2G_IDX			0
+
+
+#define HOST_COUNTER_SIZE 5
+#define NPU_MAX_TYPE 3
+
+#define HIF_TXP_SIZE (11 * 4)
+
+typedef enum {
+	P0_ = 0,
+	P1_,
+	P0__P0,
+	P0__P1
+} PortType_t;
+
+union _2_hostapdt_vend_specific{
+	struct {
+		u32 more			: 1;
+		u32 hw_rro			: 1;
+		u32 ip_frag			: 1;
+		u32 old_pkt			: 1;
+		u32 drop			: 1;
+		u32 len				: 12;
+		u32 eth_hdr_ofst	: 7;
+		u32 repeat_pkt			: 1;
+		u32 wifi_frag			: 1;
+		u32 pn_chk_fail		: 1;
+		u32 rro_v31_token	: 1;
+		u32 is_evt			: 1;
+		u32 resv			: 3;
+	}rx_info;
+	u32 ctrl_word;
+};
+
+typedef union {
+	struct {
+		unsigned char ring_num		: 1;
+		unsigned char qos_tid		: 4;
+		unsigned char wcid_dummy	: 3;//Do no use since area not enough
+	} _detail;
+	unsigned char ctrl_word;
+} npu_hif_v2_info;
+
+#define ra_dma_addr_t					unsigned long long
+typedef	ra_dma_addr_t			NDIS_PHYSICAL_ADDRESS;
+typedef void				*PNDIS_PACKET;
+
+void wifi_offload_init_npu(struct pci_dev *pdev,struct mt7996_dev *dev,int flag);
+void npu_init_rxD(struct mt7996_dev *dev,struct pci_dev *pdev);
+
+
+void get_npu_rxd(struct mt7996_dev *dev,struct pci_dev *pdev);
+int npu_enable(void);
+void set_npu_enable(int enable);
+
+int mt7996_npu_count(struct seq_file *s, void *data);
+
+void npu_mt76_dma_tx_queue_skb(struct mt76_phy *phy, struct mt76_dev *dev, struct mt76_txwi_cache *t, 
+	struct sk_buff *skb, void *txwi_ptr, struct mt76_wcid *wcid, struct ieee80211_sta *sta);
+void npu_set_rx_event(struct mt7996_dev *dev,struct pci_dev *pdev);
+void get_npu_txd(struct mt7996_dev *dev,struct pci_dev *pdev);
+void npu_mt76_init_done(struct mt7996_dev *dev,struct pci_dev *pdev);
+void npu_rro_reset_sessions(struct mt7996_dev *dev, u16 seid);
+
+extern u32 npu_wifi_offload_get_ba_node_addr(void);
+extern int (*toHostadptPktHandle_hook)(unsigned char band, struct sk_buff *skb, unsigned char *vnd, unsigned short vnd_len);
+extern struct sk_buff *(*fromHostadptPktHandle_hook)(unsigned int ringIdx, unsigned char* preschedule, unsigned int* vnd_spec);
+extern void (*hostdapt_enable_int_hook)(unsigned int ringIdx);
+extern void (*hostdapt_disable_int_hook)(unsigned int ringIdx);
+extern void (*hostdapt_registe_wifitask_hook)(unsigned int ringIdx, void *func);
+#endif
+#endif
--- a/mt7996/dma.c
+++ b/mt7996/dma.c
@@ -6,6 +6,9 @@
 #include "mt7996.h"
 #include "../dma.h"
 #include "mac.h"
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD
+#include "mt76_npu_offload.h"
+#endif
 
 int mt7996_init_tx_queues(struct mt7996_phy *phy, int idx, int n_desc,
 			  int ring_base, struct mtk_wed_device *wed)
@@ -545,8 +548,9 @@ int mt7996_dma_rro_init(struct mt7996_de
 			mt76_queue_rx_init(dev, MT_RXQ_MSDU_PAGE_BAND1, mt76_dma_rx_poll);
 			mt76_queue_rx_init(dev, MT_RXQ_MSDU_PAGE_BAND2, mt76_dma_rx_poll);
 		}
-		mt76_queue_rx_init(dev, MT_RXQ_RRO_BAND0, mt76_dma_rx_poll);
-		mt76_queue_rx_init(dev, MT_RXQ_RRO_BAND1, mt76_dma_rx_poll);
+		
+		//mt76_queue_rx_init(dev, MT_RXQ_RRO_BAND0, mt76_dma_rx_poll);
+		//mt76_queue_rx_init(dev, MT_RXQ_RRO_BAND1, mt76_dma_rx_poll);
 		if (mt7996_band_valid(dev, MT_BAND2))
 			mt76_queue_rx_init(dev, MT_RXQ_RRO_BAND2, mt76_dma_rx_poll);
 		mt76_queue_rx_init(dev, MT_RXQ_RRO_IND, mt76_dma_rx_poll);
@@ -797,6 +801,30 @@ int mt7996_dma_init(struct mt7996_dev *d
 	return 0;
 }
 
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD
+extern int tasklet_schedule_rx_data_done_npu(void);
+extern void mt7996_npu_rx_tasklet(struct tasklet_struct *t);
+extern struct mt7996_dev *MT7996;
+
+void register_rx_data_done_task(struct mt7996_dev *dev, int enable)
+{
+	if(enable)		
+	{
+		if(hostdapt_registe_wifitask_hook)
+			hostdapt_registe_wifitask_hook(0, tasklet_schedule_rx_data_done_npu);
+		tasklet_setup(&dev->npu_rx_tasklet, mt7996_npu_rx_tasklet);
+		MT7996=dev;
+	}
+	else
+	{
+		if(hostdapt_registe_wifitask_hook)
+			hostdapt_registe_wifitask_hook(0, NULL);
+		tasklet_kill(&dev->npu_rx_tasklet);
+		MT7996=NULL;
+	}	
+}
+#endif
+
 void mt7996_dma_reset(struct mt7996_dev *dev, bool force)
 {
 	struct mt76_phy *phy2 = dev->mt76.phys[MT_BAND1];
@@ -840,6 +868,10 @@ void mt7996_dma_reset(struct mt7996_dev
 	if (force)
 		mt7996_wfsys_reset(dev);
 
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD
+	if(npu_enable())
+		register_rx_data_done_task(dev, 0);
+#endif	
 	if (dev->hif2 && mtk_wed_device_active(&dev->mt76.mmio.wed_hif2))
 		mtk_wed_device_dma_reset(&dev->mt76.mmio.wed_hif2);
 
@@ -885,6 +917,12 @@ void mt7996_dma_reset(struct mt7996_dev
 		mt76_queue_rx_reset(dev, i);
 	}
 
+#if 0//def TCSUPPORT_NPU_WIFI_OFFLOAD
+	if(npu_enable()){	
+		npu_init_rxD(dev,NULL);
+		get_npu_rxd(dev,NULL);
+	}
+#endif
 	mt7996_dma_enable(dev, !force);
 }
 
--- a/mt7996/init.c
+++ b/mt7996/init.c
@@ -13,6 +13,9 @@
 #include "mcu.h"
 #include "coredump.h"
 #include "eeprom.h"
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD
+#include "mt76_npu_offload.h"
+#endif
 
 static const struct ieee80211_iface_limit if_limits_global = {
 	.max = MT7996_MAX_INTERFACES * MT7996_MAX_RADIOS,
@@ -865,6 +868,9 @@ static int mt7996_wed_rro_init(struct mt
 	struct mt7996_wed_rro_addr *addr;
 	void *ptr;
 	int i;
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD
+	int isMailBoxSuccess=0;
+#endif	
 
 	if (!dev->has_rro)
 		return 0;
@@ -899,6 +905,10 @@ static int mt7996_wed_rro_init(struct mt
 			addr->signature = 0xff;
 			addr++;
 		}
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD
+		isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_INODE_TXRX_REG_ADDR(0,
+			i, dev->wed_rro.addr_elem[i].phy_addr, 0, 0);
+#endif			
 #ifdef CONFIG_NET_MEDIATEK_SOC_WED
 		if (mtk_wed_device_active(wed) && mtk_wed_get_rx_capa(wed))
 			wed->wlan.ind_cmd.addr_elem_phys[i] = dev->wed_rro.addr_elem[i].phy_addr;
@@ -930,6 +940,10 @@ static int mt7996_wed_rro_init(struct mt
 		addr++;
 	}
 
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD
+	isMailBoxSuccess = WIFI_MAIL_API_SET_WAIT_INODE_TXRX_REG_ADDR(1,
+		0, dev->wed_rro.session.phy_addr, 0, 0);
+#endif
 	/* rro ind cmd queue init */
 	mt7996_rro_hw_init(dev);
 
@@ -998,12 +1012,15 @@ static void mt7996_wed_rro_work(struct w
 
 	while (!list_empty(&list)) {
 		struct mt7996_wed_rro_session_id *e;
-		int i;
 
 		e = list_first_entry(&list, struct mt7996_wed_rro_session_id,
 				     list);
 		list_del_init(&e->list);
-
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD
+		npu_rro_reset_sessions(dev, e->id);
+		mt7996_mcu_wed_rro_reset_sessions(dev, e->id);
+#else
+		int i;
 		for (i = 0; i < MT7996_RRO_WINDOW_MAX_LEN; i++) {
 			void *ptr = dev->wed_rro.session.ptr;
 			struct mt7996_wed_rro_addr *elem;
@@ -1026,6 +1043,7 @@ reset:
 		}
 		mt7996_mcu_wed_rro_reset_sessions(dev, e->id);
 out:
+#endif	
 		kfree(e);
 	}
 }
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -10,6 +10,9 @@
 #include "../dma.h"
 #include "mac.h"
 #include "mcu.h"
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD
+extern void register_rx_data_done_task(struct mt7996_dev *dev, int enable);
+#endif
 
 #define to_rssi(field, rcpi)	((FIELD_GET(field, rcpi) - 220) / 2)
 
@@ -455,6 +458,10 @@ mt7996_mac_fill_rx(struct mt7996_dev *de
 
 	band_idx = FIELD_GET(MT_RXD1_NORMAL_BAND_IDX, rxd1);
 	mphy = dev->mt76.phys[band_idx];
+	if(!mphy){
+//		printk("=======mt7996_mac_fill_rx!!! band_idx = %d =====\n", band_idx);	
+		return -EINVAL;
+	}	
 	phy = mphy->priv;
 	status->phy_idx = mphy->band_idx;
 
@@ -1106,6 +1113,41 @@ out:
 	mt76_put_txwi(mdev, t);
 }
 
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD
+#ifdef TCSUPPORT_NPU_WIFI_TX
+void npu_mt7996_txwi_free(struct mt76_dev *dev, struct mt76_txwi_cache *t,
+		 struct ieee80211_sta *sta, struct mt76_wcid *wcid,
+		 struct list_head *free_list)
+{
+	struct mt76_dev *mdev = dev;
+	__le32 *txwi;
+	u16 wcid_idx;
+
+	mt76_connac_txp_skb_unmap(mdev, t);
+	if (!t->skb)
+		goto out;
+
+	txwi = (__le32 *)mt76_get_txwi_ptr(mdev, t);
+	if (sta) {
+		wcid_idx = wcid->idx;
+
+		if (likely(t->skb->protocol != cpu_to_be16(ETH_P_PAE)))
+			mt7996_tx_check_aggr(sta, t->skb, wcid);
+	} else {
+		wcid_idx = le32_get_bits(txwi[9], MT_TXD9_WLAN_IDX);
+	}
+
+	__mt76_tx_complete_skb(mdev, wcid_idx, t->skb, free_list);
+
+out:
+	t->skb = NULL;
+	mt76_put_txwi(mdev, t);
+}
+
+EXPORT_SYMBOL_GPL(npu_mt7996_txwi_free);
+#endif
+#endif
+
 static void
 mt7996_mac_tx_free(struct mt7996_dev *dev, void *data, int len)
 {
@@ -2378,6 +2420,10 @@ void mt7996_mac_reset_work(struct work_s
 
 	mt7996_update_beacons(dev);
 
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD	
+	register_rx_data_done_task(dev, 1);
+#endif
+
 	ieee80211_queue_delayed_work(mt76_hw(dev), &dev->mphy.mac_work,
 				     MT7996_WATCHDOG_TIME);
 	if (phy2)
--- a/mt7996/mmio.c
+++ b/mt7996/mmio.c
@@ -13,6 +13,10 @@
 #include "mcu.h"
 #include "../trace.h"
 #include "../dma.h"
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD
+#include "mt76_npu_offload.h"
+struct mt7996_dev *MT7996;
+#endif
 
 static bool wed_enable;
 module_param(wed_enable, bool, 0644);
@@ -537,10 +541,45 @@ static void mt7996_rx_poll_complete(stru
 				    enum mt76_rxq_id q)
 {
 	struct mt7996_dev *dev = container_of(mdev, struct mt7996_dev, mt76);
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD
+	if(npu_enable())
+	{
+		if(q != MT_RXQ_RRO_IND)//wifi host not rcv data pkts after enable npu
+			mt7996_irq_enable(dev, MT_INT_RX(q)); //enable wifi irq
 
+		if(q == MT_RXQ_RRO_IND && hostdapt_enable_int_hook){
+			hostdapt_enable_int_hook(0);
+		}
+	}
+	else
+		mt7996_irq_enable(dev, MT_INT_RX(q));
+#else
 	mt7996_irq_enable(dev, MT_INT_RX(q));
+#endif
 }
 
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD
+int tasklet_schedule_rx_data_done_npu(void)
+{	  
+	struct mt7996_dev *dev;		
+   	if(MT7996){    		
+   		dev=MT7996;				
+		tasklet_schedule(&dev->npu_rx_tasklet);	
+	}				
+	return 0;
+}
+void mt7996_npu_rx_tasklet(struct tasklet_struct *t)
+{
+
+	struct mt7996_dev *dev;		
+   	if(MT7996){    		
+   		dev=MT7996;				
+		napi_schedule(&dev->mt76.napi[MT_RXQ_RRO_IND]);
+   	}
+   	
+}
+#endif
+
 /* TODO: support 2/4/6/8 MSI-X vectors */
 static void mt7996_irq_tasklet(struct tasklet_struct *t)
 {
@@ -594,8 +633,18 @@ static void mt7996_irq_tasklet(struct ta
 		napi_schedule(&dev->mt76.tx_napi);
 
 	for (i = 0; i < __MT_RXQ_MAX; i++) {
-		if ((intr & MT_INT_RX(i)))
+		if ((intr & MT_INT_RX(i))){
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD
+		if(npu_enable()){
+			if (i != MT_RXQ_MAIN_WA)
+				napi_schedule(&dev->mt76.napi[i]);
+		}
+		else
 			napi_schedule(&dev->mt76.napi[i]);
+#else
+			napi_schedule(&dev->mt76.napi[i]);
+#endif
+		}
 	}
 
 	if (intr & MT_INT_MCU_CMD) {
@@ -661,6 +710,11 @@ struct mt7996_dev *mt7996_mmio_probe(str
 		.set_channel = mt7996_set_channel,
 		.vif_link_add = mt7996_vif_link_add,
 		.vif_link_remove = mt7996_vif_link_remove,
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD	
+#ifdef TCSUPPORT_NPU_WIFI_TX
+		.npu_txwi_free = npu_mt7996_txwi_free,
+#endif
+#endif		
 	};
 	struct mt7996_dev *dev;
 	struct mt76_dev *mdev;
@@ -679,7 +733,16 @@ struct mt7996_dev *mt7996_mmio_probe(str
 	tasklet_setup(&mdev->irq_tasklet, mt7996_irq_tasklet);
 
 	mt76_wr(dev, MT_INT_MASK_CSR, 0);
-
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD	
+	if(npu_enable()){	
+		if(hostdapt_registe_wifitask_hook){
+			hostdapt_registe_wifitask_hook(0, tasklet_schedule_rx_data_done_npu);
+		}
+		tasklet_setup(&dev->npu_rx_tasklet, mt7996_npu_rx_tasklet);
+		
+		MT7996=dev; 		
+	}	
+#endif
 	return dev;
 
 error:
--- a/mt7996/mt7996.h
+++ b/mt7996/mt7996.h
@@ -456,6 +456,10 @@ struct mt7996_dev {
 		u8 type:4;
 		u8 fem:4;
 	} var;
+
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD
+	struct tasklet_struct npu_rx_tasklet;
+#endif	
 };
 
 enum {
@@ -785,6 +789,12 @@ int mt7996_tx_prepare_skb(struct mt76_de
 void mt7996_tx_token_put(struct mt7996_dev *dev);
 void mt7996_queue_rx_skb(struct mt76_dev *mdev, enum mt76_rxq_id q,
 			 struct sk_buff *skb, u32 *info);
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD
+#ifdef TCSUPPORT_NPU_WIFI_TX
+void npu_mt7996_txwi_free(struct mt76_dev *dev, struct mt76_txwi_cache *t,
+		 struct ieee80211_sta *sta, struct mt76_wcid *wcid,	 struct list_head *free_list);
+#endif
+#endif
 void mt7996_rx_token_put(struct mt7996_dev *dev);
 void mt7996_rro_msdu_pg_free(struct mt7996_dev *dev);
 void mt7996_rro_rx_process(struct mt76_dev *mdev, void *data);
--- a/mt7996/pci.c
+++ b/mt7996/pci.c
@@ -9,8 +9,11 @@
 #include "mt7996.h"
 #include "mac.h"
 #include "../trace.h"
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD
+#include "mt76_npu_offload.h"
+#endif
 
-static bool rro_enable = false;
+static bool rro_enable = true;
 module_param(rro_enable, bool, 0644);
 
 static LIST_HEAD(hif_list);
@@ -194,7 +197,22 @@ static int mt7996_pci_probe(struct pci_d
 	ret = mt7996_register_device(dev);
 	if (ret)
 		goto free_hif2_irq;
-
+#ifdef TCSUPPORT_NPU_WIFI_OFFLOAD
+	if(npu_enable()){			
+		wifi_offload_init_npu(pdev,dev,0);
+		npu_init_rxD(dev,pdev);
+		get_npu_rxd(dev,pdev);
+		
+		mt7996_irq_disable(dev, MT_INT_RRO_RX_DONE);
+#ifdef TCSUPPORT_NPU_WIFI_TX
+		get_npu_txd(dev,pdev);
+		npu_set_rx_event(dev,pdev);
+
+		mt7996_irq_disable(dev, MT_INT_RX(MT_RXQ_MAIN_WA));
+#endif
+		npu_mt76_init_done(dev,pdev);
+	}
+#endif
 	return 0;
 
 free_hif2_irq:
