--- a/npu.c
+++ b/npu.c
@@ -201,7 +201,8 @@ void get_npu_rxd(struct mt7996_dev *dev,
 	if(isMailBoxSuccess != 1) 
 		printk("Error: NPU_WIFI_OFFLOAD get ring1 base fail\n");
 	npu_alloc_pa=(NDIS_PHYSICAL_ADDRESS)get_RingBase.info;
-	writel(npu_alloc_pa,&dev->mt76.q_rx[MT_RXQ_RRO_BAND1].regs->desc_base);		
+	if (dev->mt76.q_rx[MT_RXQ_RRO_BAND1].regs)
+		writel(npu_alloc_pa,&dev->mt76.q_rx[MT_RXQ_RRO_BAND1].regs->desc_base);		
 
 #if MT76_RROV31
 		/*rxdmad_c ring*/
@@ -209,7 +210,8 @@ void get_npu_rxd(struct mt7996_dev *dev,
 		if (isMailBoxSuccess != 1)
 			printk("%s:: %d:: Error: WIFI_MAIL_API_GET_WAIT_RXDESC_BASE fail\n",__func__, __LINE__);
 		npu_alloc_pa=(NDIS_PHYSICAL_ADDRESS)get_RingBase.info;
-		writel(npu_alloc_pa,&dev->mt76.q_rx[MT_RXQ_RRO_RXDMAD_C].regs->desc_base);	
+		if (dev->mt76.q_rx[MT_RXQ_RRO_RXDMAD_C].regs)
+			writel(npu_alloc_pa,&dev->mt76.q_rx[MT_RXQ_RRO_RXDMAD_C].regs->desc_base);	
 	
 #else
 	/* MSDU PG ring */
--- a/dma.c
+++ b/dma.c
@@ -1101,9 +1101,11 @@ __mt76_dma_init(struct mt76_dev *dev, en
 	int i;
 
 	if (qid < __MT_RXQ_MAX && dev->q_rx[qid].ndesc) {
-		netif_napi_add(&dev->napi_dev, &dev->napi[qid], poll);
-		mt76_dma_rx_fill(dev, &dev->q_rx[qid], false);
-		napi_enable(&dev->napi[qid]);
+		if (!test_bit(NAPI_STATE_LISTED, &dev->napi[qid].state)) {
+			netif_napi_add(&dev->napi_dev, &dev->napi[qid], poll);
+			mt76_dma_rx_fill(dev, &dev->q_rx[qid], false);
+			napi_enable(&dev->napi[qid]);
+		}
 		return 0;
 	}
 
