--- a/npu.c
+++ b/npu.c
@@ -1,3 +1,5 @@
+#include "mt76.h"
+#include "dma.h"
 #include "npu.h"
 
 #ifdef CONFIG_NET_AIROHA_NPU
@@ -311,39 +313,6 @@ void get_npu_txd(struct mt7996_dev *dev,
 }
 EXPORT_SYMBOL_GPL(get_npu_txd);
 
-char npu_mt76_token_release(struct mt76_dev *dev, unsigned int token, struct mt76_wcid *wcid, struct ieee80211_sta *sta)
-{	
-	struct mt76_txwi_cache *txwi_t;	
-	struct sk_buff *skb, *tmp;
-	LIST_HEAD(free_list);
-	
-	spin_lock_bh(&dev->token_lock);
-	
-	txwi_t = idr_remove(&dev->token, token);
-	
-	if (txwi_t) {
-		struct mt76_phy *phy = mt76_dev_phy(dev, txwi_t->phy_idx);
-		
-		dev->token_count--;
-		phy->tokens--;
-	}
-	
-	spin_unlock_bh(&dev->token_lock);
-	if (txwi_t) {
-		dev->drv->npu_txwi_free(dev, txwi_t, sta, wcid, &free_list);		
-		txwi_t->jiffies = 0;
-	}
-	
-	mt76_worker_schedule(&dev->tx_worker);
-	
-	list_for_each_entry_safe(skb, tmp, &free_list, list) {
-		skb_list_del_init(skb);
-		napi_consume_skb(skb, 1);
-	}
-
-	return 0;
-}
-
 void npu_set_rx_event(struct mt7996_dev *dev,struct pci_dev *pdev)
 {
 	struct airoha_npu *npu = dev->mt76.mmio.npu;
@@ -429,12 +398,15 @@ static irqreturn_t mt76_npu_irq_handler(
 	return IRQ_HANDLED;
 }
 
-int mt76_npu_tx_queue_skb(struct mt76_phy *phy, struct sk_buff *skb,
-			  void *txwi_ptr)
+int mt76_npu_dma_add_buf(struct mt76_phy *phy, struct mt76_queue *q,
+			 struct sk_buff *skb, struct mt76_queue_buf *buf,
+			 void *txwi_ptr)
 {
+	u16 txwi_len = min_t(u16, MT_TXD_SIZE + HIF_TXP_SIZE, NPU_TXWI_LEN);
+	struct airoha_npu_tx_dma_desc *desc = q->desc;
 	struct mt76_dev *dev = phy->dev;
 	struct airoha_npu *npu;
-	int qid, ret = 0;
+	int ret;
 
 	rcu_read_lock();
 	npu = rcu_dereference(dev->mmio.npu);
@@ -443,19 +415,54 @@ int mt76_npu_tx_queue_skb(struct mt76_ph
 		goto out;
 	}
 
-	qid = phy == dev->phys[1] ? MT_BAND1 : MT_BAND0;
-	ret = npu->ops.wlan_tx_enqueue(npu, qid, skb, txwi_ptr,
-				       MT_TXD_SIZE + HIF_TXP_SIZE);
-	if (ret)
-		goto out;
+	/* FIXME: Take into account unlinear skbs */
+	memcpy(desc[q->head].txwi, txwi_ptr, txwi_len);
+	desc[q->head].addr = buf->addr;
+	desc[q->head].ctrl = FIELD_PREP(NPU_TX_DMA_DESC_VEND_LEN_MASK, txwi_len) |
+			     FIELD_PREP(NPU_TX_DMA_DESC_LEN_MASK, skb->len) |
+			     NPU_TX_DMA_DESC_DONE_MASK;
+
+	ret = q->head;
+	q->entry[q->head].skip_buf0 = true;
+	q->entry[q->head].skip_buf1 = true;
+	q->entry[q->head].txwi = NULL;
+	q->entry[q->head].skb = NULL;
+	q->entry[q->head].wcid = 0xffff;
 
-	ret = npu->ops.wlan_kick_tx_queue(npu, qid);
+	q->head = (q->head + 1) % q->ndesc;
+	q->queued++;
 out:
 	rcu_read_unlock();
 
 	return ret;
 }
 
+void mt76_npu_txdesc_cleanup(struct mt76_queue *q, int index)
+{
+	struct airoha_npu_tx_dma_desc *desc = q->desc;
+
+	if (!mt76_queue_is_npu_tx(q))
+		return;
+
+	desc[index].ctrl &= ~NPU_TX_DMA_DESC_DONE_MASK;
+}
+
+void mt76_npu_queue_setup(struct mt76_dev *dev, struct mt76_queue *q)
+{
+	int qid = FIELD_GET(MT_QFLAG_WED_RING, q->flags);
+	bool xmit = mt76_queue_is_npu_tx(q);
+	struct airoha_npu *npu;
+
+	if (!mt76_queue_is_npu(q))
+		return;
+
+	rcu_read_lock();
+	npu = rcu_dereference(dev->mmio.npu);
+	if (npu)
+		q->wed_regs = npu->ops.wlan_get_queue_addr(npu, qid, xmit);
+	rcu_read_unlock();
+}
+
 int mt76_npu_init(struct mt76_dev *dev)
 {
 	struct airoha_npu *npu;
@@ -479,6 +486,7 @@ int mt76_npu_init(struct mt76_dev *dev)
 	if (err)
 		goto unlock;
 
+	dev->dma_dev = npu->dev;
 	rcu_assign_pointer(dev->mmio.npu, npu);
 	synchronize_rcu();
 
--- a/dma.c
+++ b/dma.c
@@ -10,37 +10,6 @@
 #include "npu.h"
 #endif
 
-#if IS_ENABLED(CONFIG_NET_MEDIATEK_SOC_WED)
-
-#define Q_READ(_q, _field) ({						\
-	u32 _offset = offsetof(struct mt76_queue_regs, _field);		\
-	u32 _val;							\
-	if ((_q)->flags & MT_QFLAG_WED)					\
-		_val = mtk_wed_device_reg_read((_q)->wed,		\
-					       ((_q)->wed_regs +	\
-					        _offset));		\
-	else								\
-		_val = readl(&(_q)->regs->_field);			\
-	_val;								\
-})
-
-#define Q_WRITE(_q, _field, _val)	do {				\
-	u32 _offset = offsetof(struct mt76_queue_regs, _field);		\
-	if ((_q)->flags & MT_QFLAG_WED)					\
-		mtk_wed_device_reg_write((_q)->wed,			\
-					 ((_q)->wed_regs + _offset),	\
-					 _val);				\
-	else								\
-		writel(_val, &(_q)->regs->_field);			\
-} while (0)
-
-#else
-
-#define Q_READ(_q, _field)		readl(&(_q)->regs->_field)
-#define Q_WRITE(_q, _field, _val)	writel(_val, &(_q)->regs->_field)
-
-#endif
-
 static struct mt76_txwi_cache *
 mt76_alloc_txwi(struct mt76_dev *dev)
 {
@@ -198,6 +167,11 @@ mt76_dma_sync_idx(struct mt76_dev *dev,
 	else
 #endif		
 		Q_WRITE(q, ring_size, q->ndesc);
+
+	if (mt76_queue_is_npu(q)) {
+		writel(q->desc_dma, &q->regs->desc_base);
+		writel(q->ndesc, &q->regs->ring_size);
+	}
 	q->head = Q_READ(q, dma_idx);
 	q->tail = q->head;
 }
@@ -208,7 +182,8 @@ void __mt76_dma_queue_reset(struct mt76_
 	if (!q || !q->ndesc)
 		return;
 
-	if (!mt76_queue_is_wed_rro_ind(q) && !mt76_queue_is_wed_rro_rxdmad_c(q)) {   //rro3.1
+	if (!mt76_queue_is_npu(q) && !mt76_queue_is_wed_rro_ind(q) &&
+	    !mt76_queue_is_wed_rro_rxdmad_c(q)) {   //rro3.1
 		struct mt76_desc *desc = q->desc;
 		int i;
 
@@ -401,15 +376,6 @@ mt76_dma_tx_cleanup_idx(struct mt76_dev
 static void
 mt76_dma_kick_queue(struct mt76_dev *dev, struct mt76_queue *q)
 {
-#ifdef CONFIG_NET_AIROHA_NPU
-	if(airoha_npu_device_active(dev->mmio.npu)){
-		if(!(dev->phy.q_tx[0] == q ||(dev->phys[1] && dev->phys[1]->q_tx[0] == q))){
-			wmb();
-			Q_WRITE(q, cpu_idx, q->head);
-		}
-		return;        
-	}
-#endif
 	wmb();
 	Q_WRITE(q, cpu_idx, q->head);
 }
@@ -431,6 +397,7 @@ mt76_dma_tx_cleanup(struct mt76_dev *dev
 
 	while (q->queued > 0 && q->tail != last) {
 		mt76_dma_tx_cleanup_idx(dev, q, q->tail, &entry);
+		mt76_npu_txdesc_cleanup(q, q->tail);
 		mt76_queue_tx_complete(dev, q, &entry);
 
 		if (entry.txwi) {
@@ -693,6 +660,10 @@ mt76_dma_tx_queue_skb(struct mt76_phy *p
 	dma_addr_t addr;
 	u8 *txwi;
 
+	/* FIXME: Take into account unlinear skbs */
+	if (mt76_npu_device_active(dev) && skb_linearize(skb))
+		goto free_skb;
+
 	if (test_bit(MT76_RESET, &phy->state)) {
 		phy->tx_dbg_stats.tx_drop[MT_TX_DROP_RESET_STATE]++;
 		goto free_skb;
@@ -755,18 +726,11 @@ mt76_dma_tx_queue_skb(struct mt76_phy *p
 		goto unmap;
 
 	phy->tx_dbg_stats.tx_to_hw++;
-#ifdef CONFIG_NET_AIROHA_NPU
-	if(airoha_npu_device_active(dev->mmio.npu)) {
-		if (mt76_npu_tx_queue_skb(phy, skb, txwi))
-			dev_kfree_skb(skb);
-		return -1;
-    }else
-    	return mt76_dma_add_buf(dev, q, tx_info.buf, tx_info.nbuf,
-				tx_info.info, tx_info.skb, t);
-#else	
+	if (mt76_npu_device_active(dev))
+		return mt76_npu_dma_add_buf(phy, q, skb, &tx_info.buf[1], txwi);
+
 	return mt76_dma_add_buf(dev, q, tx_info.buf, tx_info.nbuf,
 				tx_info.info, tx_info.skb, t);
-#endif
 
 unmap:
 	for (n--; n > 0; n--)
@@ -861,9 +825,15 @@ mt76_dma_alloc_queue(struct mt76_dev *de
 	q->ndesc = n_desc;
 	q->buf_size = bufsize;
 	q->hw_idx = idx;
+	q->dev = dev;
+
+	if (mt76_queue_is_wed_rro_ind(q))
+		size = sizeof(struct mt76_wed_rro_desc);
+	else if (mt76_queue_is_npu_tx(q))
+		size = sizeof(struct airoha_npu_tx_dma_desc);
+	else
+		size = sizeof(struct mt76_desc);
 
-	size = mt76_queue_is_wed_rro_ind(q) ? sizeof(struct mt76_wed_rro_desc)
-					    : sizeof(struct mt76_desc);
 	q->desc = dmam_alloc_coherent(dev->dma_dev, q->ndesc * size,
 				      &q->desc_dma, GFP_KERNEL);
 	if (!q->desc)
@@ -891,6 +861,7 @@ mt76_dma_alloc_queue(struct mt76_dev *de
 	if (!q->entry)
 		return -ENOMEM;
 
+	mt76_npu_queue_setup(dev, q);
 	ret = mt76_wed_dma_setup(dev, q, false);
 	if (ret)
 		return ret;
--- a/mt76.h
+++ b/mt76.h
@@ -63,6 +63,7 @@
 #define MT_QFLAG_WED		BIT(5)
 #define MT_QFLAG_WED_RRO	BIT(6)
 #define MT_QFLAG_WED_RRO_EN	BIT(7)
+#define MT_QFLAG_NPU		BIT(8)
 
 #define __MT_WED_Q(_type, _n)	(MT_QFLAG_WED | \
 				 FIELD_PREP(MT_QFLAG_WED_TYPE, _type) | \
@@ -78,6 +79,12 @@
 /*rro3.1*/
 #define MT_WED_RRO_Q_RXDMAD_C	__MT_WED_RRO_Q(MT76_WED_RRO_Q_RXDMAD_C, 0)
 
+#define __MT_NPU_Q(_type, _n)	(MT_QFLAG_NPU | \
+				 FIELD_PREP(MT_QFLAG_WED_TYPE, _type) | \
+				 FIELD_PREP(MT_QFLAG_WED_RING, _n))
+#define MT_NPU_Q_TX(_n)		__MT_NPU_Q(MT76_WED_Q_TX, _n)
+#define MT_NPU_Q_RX(_n)		__MT_NPU_Q(MT76_WED_Q_RX, _n)
+
 #define AMPDU_ADDBA_SUCC_SHFT IEEE80211_NUM_TIDS
 
 struct mt76_dev;
@@ -297,6 +304,7 @@ struct mt76_queue {
 	u8 magic_cnt;
 
 	struct mtk_wed_device *wed;
+	struct mt76_dev *dev;
 	u32 wed_regs;
 
 	dma_addr_t desc_dma;
@@ -676,11 +684,6 @@ struct mt76_driver_ops {
 
 	void (*sta_remove)(struct mt76_dev *dev, struct ieee80211_vif *vif,
 			   struct ieee80211_sta *sta);
-#ifdef CONFIG_NET_AIROHA_NPU
-	void (*npu_txwi_free)(struct mt76_dev *dev, struct mt76_txwi_cache *t,
-			 struct ieee80211_sta *sta, struct mt76_wcid *wcid,
-			 struct list_head *free_list);
-#endif
 };
 
 struct mt76_channel_state {
@@ -1771,10 +1774,18 @@ int mt76_testmode_set_state(struct mt76_
 int mt76_testmode_alloc_skb(struct mt76_phy *phy, u32 len);
 
 int mt76_npu_rx_poll(struct napi_struct *napi, int budget);
-int mt76_npu_tx_queue_skb(struct mt76_phy *phy, struct sk_buff *skb,
-			  void *txwi_ptr);
+int mt76_npu_dma_add_buf(struct mt76_phy *phy, struct mt76_queue *q,
+			 struct sk_buff *skb, struct mt76_queue_buf *buf,
+			 void *txwi_ptr);
 int mt76_npu_init(struct mt76_dev *dev);
 void mt76_npu_deinit(struct mt76_dev *dev);
+void mt76_npu_queue_setup(struct mt76_dev *dev, struct mt76_queue *q);
+void mt76_npu_txdesc_cleanup(struct mt76_queue *q, int index);
+
+static inline bool mt76_npu_device_active(struct mt76_dev *dev)
+{
+	return !!rcu_access_pointer(dev->mmio.npu);
+}
 
 static inline void
 mt76_testmode_param_set(struct mt76_testmode_data *td, u16 idx)
@@ -2021,6 +2032,23 @@ static inline bool mt76_queue_is_wed_rx(
 	       FIELD_GET(MT_QFLAG_WED_TYPE, q->flags) == MT76_WED_Q_RX;
 }
 
+static inline bool mt76_queue_is_npu(struct mt76_queue *q)
+{
+	return q->flags & MT_QFLAG_NPU;
+}
+
+static inline bool mt76_queue_is_npu_tx(struct mt76_queue *q)
+{
+	return mt76_queue_is_npu(q) &&
+	       FIELD_GET(MT_QFLAG_WED_TYPE, q->flags) == MT76_WED_Q_TX;
+}
+
+static inline bool mt76_queue_is_npu_rx(struct mt76_queue *q)
+{
+	return mt76_queue_is_npu(q) &&
+	       FIELD_GET(MT_QFLAG_WED_TYPE, q->flags) == MT76_WED_Q_RX;
+}
+
 struct mt76_txwi_cache *
 mt76_token_release(struct mt76_dev *dev, int token, bool *wake);
 int mt76_token_consume(struct mt76_dev *dev, struct mt76_txwi_cache **ptxwi,
--- a/dma.h
+++ b/dma.h
@@ -45,6 +45,73 @@
 #define MT_FCE_INFO_LEN			4
 #define MT_RX_RXWI_LEN			32
 
+#if IS_ENABLED(CONFIG_NET_MEDIATEK_SOC_WED)
+
+#define Q_READ(_q, _field) ({						\
+	u32 _offset = offsetof(struct mt76_queue_regs, _field);		\
+	u32 _val;							\
+	if ((_q)->flags & MT_QFLAG_WED)					\
+		_val = mtk_wed_device_reg_read((_q)->wed,		\
+					       ((_q)->wed_regs +	\
+					        _offset));		\
+	else								\
+		_val = readl(&(_q)->regs->_field);			\
+	_val;								\
+})
+
+#define Q_WRITE(_q, _field, _val)	do {				\
+	u32 _offset = offsetof(struct mt76_queue_regs, _field);		\
+	if ((_q)->flags & MT_QFLAG_WED)					\
+		mtk_wed_device_reg_write((_q)->wed,			\
+					 ((_q)->wed_regs + _offset),	\
+					 _val);				\
+	else								\
+		writel(_val, &(_q)->regs->_field);			\
+} while (0)
+
+#elif (IS_BUILTIN(CONFIG_NET_AIROHA_NPU) || IS_MODULE(CONFIG_NET_AIROHA_NPU))
+
+#define Q_READ(_q, _field) ({						\
+	u32 _offset = offsetof(struct mt76_queue_regs, _field);		\
+	u32 _val;							\
+	if ((_q)->flags & MT_QFLAG_NPU) {				\
+		struct airoha_npu *npu;					\
+									\
+		rcu_read_lock();					\
+		npu = rcu_dereference(q->dev->mmio.npu);		\
+		if (npu)						\
+			regmap_read(npu->regmap,			\
+				    ((_q)->wed_regs + _offset), &_val);	\
+		rcu_read_unlock();					\
+	} else {							\
+		_val = readl(&(_q)->regs->_field);			\
+	}								\
+	_val;								\
+})
+
+#define Q_WRITE(_q, _field, _val)	do {				\
+	u32 _offset = offsetof(struct mt76_queue_regs, _field);		\
+	if ((_q)->flags & MT_QFLAG_NPU) {				\
+		struct airoha_npu *npu;					\
+									\
+		rcu_read_lock();					\
+		npu = rcu_dereference(q->dev->mmio.npu);		\
+		if (npu)						\
+			regmap_write(npu->regmap,			\
+				     ((_q)->wed_regs + _offset), _val);	\
+		rcu_read_unlock();					\
+	} else {							\
+		writel(_val, &(_q)->regs->_field);			\
+	}								\
+} while (0)
+
+#else
+
+#define Q_READ(_q, _field)		readl(&(_q)->regs->_field)
+#define Q_WRITE(_q, _field, _val)	writel(_val, &(_q)->regs->_field)
+
+#endif
+
 struct mt76_desc {
 	__le32 buf0;
 	__le32 ctrl;
--- a/mt7996/dma.c
+++ b/mt7996/dma.c
@@ -33,6 +33,10 @@ int mt7996_init_tx_queues(struct mt7996_
 			flags = MT_WED_Q_TX(idx);
 	}
 #endif
+
+	if (mt76_npu_device_active(&dev->mt76))
+		flags = MT_NPU_Q_TX(phy->mt76->band_idx);
+
 	return mt76_connac_init_tx_queues(phy->mt76, idx, n_desc,
 					  ring_base, wed, flags);
 }
--- a/mt7996/mmio.c
+++ b/mt7996/mmio.c
@@ -892,10 +892,7 @@ struct mt7996_dev *mt7996_mmio_probe(str
 		.sta_event = mt7996_mac_sta_event,
 		.sta_remove = mt7996_mac_sta_remove,
 		.update_survey = mt7996_update_channel,
-		.set_channel = mt7996_set_channel,
-#ifdef CONFIG_NET_AIROHA_NPU	
-		.npu_txwi_free = npu_mt7996_txwi_free,
-#endif			
+		.set_channel = mt7996_set_channel,	
 	};
 	struct mt7996_dev *dev;
 	struct mt76_dev *mdev;
